{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "complex-berkeley",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "radical-anxiety",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import csv\n",
    "import os\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "import math\n",
    "import altair as alt\n",
    "import glob\n",
    "import sys\n",
    "\n",
    "module_path = os.path.abspath(os.path.join('..', 'scripts'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "#end if\n",
    "\n",
    "import utils\n",
    "from utils import read_json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "decreased-tract",
   "metadata": {},
   "source": [
    "## Declares"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "outer-supplier",
   "metadata": {},
   "outputs": [],
   "source": [
    "TAG = \"mammalian_REM2\"\n",
    "DATA_DIRECTORY = os.path.join(\"..\", \"results\", TAG)\n",
    "\n",
    "JSONS_BS = glob.glob(os.path.join(DATA_DIRECTORY, \"*.BUSTEDS.json\"))\n",
    "JSONS_BSMH = glob.glob(os.path.join(DATA_DIRECTORY, \"*.BUSTEDS+MH.json\"))\n",
    "JSONS_BASE = glob.glob(os.path.join(DATA_DIRECTORY, \"*.BUSTED.json\"))\n",
    "JSONS_BMH = glob.glob(os.path.join(DATA_DIRECTORY, \"*.BUSTED+MH.json\"))\n",
    "\n",
    "# Create tables folder\n",
    "OUTPUT_CSV = TAG + \"_BUSTED_ModelTest.csv\"\n",
    "\n",
    "ER_Threshold = 10\n",
    "ER_Threshold_loose = 1\n",
    "\n",
    "pval_Threshold = 0.05\n",
    "\n",
    "Tests = 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "sealed-flexibility",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define function to calculate cv\n",
    "#cv = lambda x: np.std(x, ddof=1) / np.mean(x) * 100 \n",
    "cv = lambda x: np.std(x) / np.mean(x)\n",
    "pctchg = lambda a, b: (a / b) * 100\n",
    "\n",
    "\"\"\"\n",
    "def read_json(filename):\n",
    "    if os.stat(filename).st_size == 0: \n",
    "        print(\"# -- Error -- file is empty:\", filename)\n",
    "        return []\n",
    "    #end if\n",
    "    with open(filename, \"r\") as fh:\n",
    "        json_data = json.load(fh)\n",
    "    fh.close()\n",
    "    return json_data\n",
    "#end method\n",
    "\"\"\"\n",
    "\n",
    "def convergence_issue(json_data):\n",
    "    # For example if logL for a more complex model is LOWER than it is for a less complex model.\n",
    "    convergence_issue = False \n",
    "    return \"\"\n",
    "    \"\"\"\n",
    "    Constrained model\n",
    "    MG94xREV with separate rates for branch sets\n",
    "    Nucleotide GTR\n",
    "    Unconstrained model\n",
    "    \"\"\"\n",
    "    \n",
    "    _models = [\"Nucleotide GTR\", \n",
    "                    \"MG94xREV with separate rates for branch sets\", \n",
    "                    \"Constrained model\", \n",
    "                    \"Unconstrained model\"]\n",
    "    model_LL = {}\n",
    "    \n",
    "    for _model in _models:\n",
    "        print(_model)\n",
    "        _LL = json_data[\"fits\"][_model][\"Log Likelihood\"]\n",
    "\n",
    "        \n",
    "    \"\"\"\n",
    "    return\n",
    "    #Nucleotide_GTR = BS_data[\"fits\"][\"Nucleotide GTR\"].get(\"Log Likelihood\", 10)\n",
    "    #MG94_REV = BS_data[\"fits\"][\"MG94xREV with separate rates for branch sets\"].get(\"Log Likelihood\",10)\n",
    "    try:\n",
    "        Constrained_model = BS_data[\"fits\"][\"Constrained model\"].get(\"Log Likelihood\", 10)\n",
    "    except:\n",
    "        Constrained_model = 10\n",
    "    #end try\n",
    "    Unconstrained_model = BS_data[\"fits\"][\"Unconstrained model\"].get(\"Log Likelihood\", 10)\n",
    "    if Nucleotide_GTR < MG94_REV and MG94_REV < Constrained_model and Constrained_model < Unconstrained_model:\n",
    "        pass\n",
    "    else:\n",
    "        issue = True\n",
    "    #end if\n",
    "    if issue == False:\n",
    "        return \"\"\n",
    "    else:\n",
    "        return \"ISSUE DETECTED\"\n",
    "    #end if\n",
    "    \"\"\"\n",
    "#end method\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def convergence_issue_old(BS_data):\n",
    "    # For example if logL for a more complex model is LOWER than it is for a less complex model.\n",
    "    convergence_issue = False \n",
    "    \n",
    "    #Nucleotide_GTR = BS_data[\"fits\"][\"Nucleotide GTR\"].get(\"Log Likelihood\", 10)\n",
    "    #MG94_REV = BS_data[\"fits\"][\"MG94xREV with separate rates for branch sets\"].get(\"Log Likelihood\",10)\n",
    "    \n",
    "    try:\n",
    "        Constrained_model = BS_data[\"fits\"][\"Constrained model\"].get(\"Log Likelihood\", 10)\n",
    "    except:\n",
    "        Constrained_model = 10\n",
    "    #end try\n",
    "    \n",
    "    Unconstrained_model = BS_data[\"fits\"][\"Unconstrained model\"].get(\"Log Likelihood\", 10)\n",
    "    \n",
    "    if Nucleotide_GTR < MG94_REV and MG94_REV < Constrained_model and Constrained_model < Unconstrained_model:\n",
    "        pass\n",
    "    else:\n",
    "        issue = True\n",
    "    #end if\n",
    "    \n",
    "    if issue == False:\n",
    "        return \"\"\n",
    "    else:\n",
    "        return \"ISSUE DETECTED\"\n",
    "    #end if\n",
    "#end method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "similar-active",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process(FILES, fileending, method, pval_Threshold, Tests):\n",
    "    df_dict = {}\n",
    "    Bonferroni_pval = pval_Threshold / Tests\n",
    "    for item in FILES:\n",
    "        print(\"\\t Processing:\", item)\n",
    "        basename = \"\"\n",
    "        if fileending in os.path.basename(item):\n",
    "             basename = os.path.basename(item).replace(fileending, \"\").replace(\".best-gard\", \"\")\n",
    "             #basename = os.path.basename(item)\n",
    "             #print(\"\\t Basename:\", basename)\n",
    "        #end if\n",
    "        for fext in [\".phy\", \".fasta\", \".nex\", \"-align-dna.fas\", \"-Aligned-DNA.fas\"]:\n",
    "            basename = basename.replace(fext, \"\")\n",
    "        #end for\n",
    "        json_data = read_json(item) # Read json\n",
    "        if json_data == []:\n",
    "            continue # Empty file\n",
    "        #end if\n",
    "        df_dict[basename] = {\"Method\": method}\n",
    "        df_dict[basename].update({\"Sequences\": json_data[\"input\"][\"number of sequences\"]})\n",
    "        df_dict[basename].update({\"Codons\": json_data[\"input\"][\"number of sites\"]})\n",
    "        df_dict[basename].update({\"LRT p-value\": json_data[\"test results\"][\"p-value\"]})\n",
    "        df_dict[basename].update({\"Bonferroni p-value\": Bonferroni_pval})\n",
    "        Bonferroni_sig = False\n",
    "        if json_data[\"test results\"][\"p-value\"] <= Bonferroni_pval:\n",
    "            Bonferroni_sig = True\n",
    "        #end if\n",
    "        # Bonferroni Test\n",
    "        df_dict[basename].update({\"Bonferroni significant\": str(Bonferroni_sig)})\n",
    "        # cAIC\n",
    "        df_dict[basename].update({\"cAIC\": json_data[\"fits\"][\"Unconstrained model\"][\"AIC-c\"]})\n",
    "        #Omegas and proportions\n",
    "        data = json_data[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Test\"]\n",
    "        w1 = data[\"0\"][\"omega\"]\n",
    "        p1 = data[\"0\"][\"proportion\"]\n",
    "        w2 = data[\"1\"][\"omega\"]\n",
    "        p2 = data[\"1\"][\"proportion\"]\n",
    "        w3 = data[\"2\"][\"omega\"]\n",
    "        p3 = data[\"2\"][\"proportion\"]\n",
    "        df_dict[basename].update({\"w1\": w1, \"p1\": p1})\n",
    "        df_dict[basename].update({\"w2\": w2, \"p2\": p2})\n",
    "        df_dict[basename].update({\"w3\": w3, \"p3\": p3})\n",
    "        # SRV rates and proportions\n",
    "        if method == \"BUSTEDS+MH\" or method == \"BUSTEDS\":\n",
    "            data = json_data[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Synonymous site-to-site rates\"]\n",
    "            s1 = data[\"0\"][\"rate\"]\n",
    "            s_p1 = data[\"0\"][\"proportion\"]\n",
    "            s2 = data[\"1\"][\"rate\"]\n",
    "            s_p2 = data[\"1\"][\"proportion\"]\n",
    "            s3 = data[\"2\"][\"rate\"]\n",
    "            s_p3 = data[\"2\"][\"proportion\"]\n",
    "            df_dict[basename].update({\"SRV1\": s1, \"SRV_p1\": s_p1})\n",
    "            df_dict[basename].update({\"SRV2\": s2, \"SRV_p2\": s_p2})\n",
    "            df_dict[basename].update({\"SRV3\": s3, \"SRV_p3\": s_p3})\n",
    "        #end if\n",
    "        # DH rate, TH rate, TH_SI rate\n",
    "        if method == \"BUSTEDS+MH\" or method == \"BUSTED+MH\":\n",
    "            df_dict[basename].update({\"DH_Rate\": float(json_data[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"rate at which 2 nucleotides are changed instantly within a single codon\"])})\n",
    "            df_dict[basename].update({\"TH_Rate\": float(json_data[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"rate at which 3 nucleotides are changed instantly within a single codon\"])})\n",
    "            #df_dict[basename].update({\"TH_Rate_SI\": float(json_data[\"fits\"][\"Unconstrained model\"][\"rate at which 3 nucleotides are changed instantly within a single codon between synonymous codon islands\"])})\n",
    "        #end if\n",
    "        # Convergence issues check\n",
    "        df_dict[basename].update({\"Convergence_Issue\": convergence_issue(json_data)})\n",
    "    # end for\n",
    "    df = pd.DataFrame.from_dict(df_dict, orient=\"index\")\n",
    "    df = df.reset_index()\n",
    "    df.index += 1\n",
    "    df.rename(columns={'index': 'Gene'}, inplace = True)\n",
    "    return df\n",
    "#end method\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "defa9746-0504-42cb-9b81-ca70e1ae9ad6",
   "metadata": {},
   "source": [
    "## Process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "academic-shoot",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Processing BUSTED[S] files\n",
      "\t Processing: ../results/mammalian_REM2/mammalian_REM2_codons.SA.FilterOutliers.fasta.BUSTEDS.json\n",
      "\n",
      "# Processing BUSTED[S]-MH files\n",
      "\t Processing: ../results/mammalian_REM2/mammalian_REM2_codons.SA.FilterOutliers.fasta.BUSTEDS+MH.json\n",
      "\n",
      "# Processing BUSTED files\n",
      "\t Processing: ../results/mammalian_REM2/mammalian_REM2_codons.SA.FilterOutliers.fasta.BUSTED.json\n",
      "\n",
      "# Processing BUSTED-MH files\n",
      "\t Processing: ../results/mammalian_REM2/mammalian_REM2_codons.SA.FilterOutliers.fasta.BUSTED+MH.json\n"
     ]
    }
   ],
   "source": [
    "print(\"# Processing BUSTED[S] files\")\n",
    "df_BUSTEDS = process(JSONS_BS , \".BUSTEDS.json\", \"BUSTEDS\", pval_Threshold, Tests)\n",
    "\n",
    "print(\"\\n# Processing BUSTED[S]-MH files\")\n",
    "df_BUSTEDS_MH = process(JSONS_BSMH, \".BUSTEDS+MH.json\", \"BUSTEDS+MH\", pval_Threshold, Tests)\n",
    "\n",
    "print(\"\\n# Processing BUSTED files\")\n",
    "df_BUSTED = process(JSONS_BASE , \".BUSTED.json\", \"BUSTED\", pval_Threshold, Tests)\n",
    "print(\"\\n# Processing BUSTED-MH files\")\n",
    "df_BUSTED_MH = process(JSONS_BMH, \".BUSTED+MH.json\", \"BUSTED+MH\", pval_Threshold, Tests)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5509d43",
   "metadata": {},
   "source": [
    "## Concatenate tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "908b0296",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Method</th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Codons</th>\n",
       "      <th>LRT p-value</th>\n",
       "      <th>Bonferroni p-value</th>\n",
       "      <th>Bonferroni significant</th>\n",
       "      <th>cAIC</th>\n",
       "      <th>w1</th>\n",
       "      <th>p1</th>\n",
       "      <th>...</th>\n",
       "      <th>p3</th>\n",
       "      <th>SRV1</th>\n",
       "      <th>SRV_p1</th>\n",
       "      <th>SRV2</th>\n",
       "      <th>SRV_p2</th>\n",
       "      <th>SRV3</th>\n",
       "      <th>SRV_p3</th>\n",
       "      <th>Convergence_Issue</th>\n",
       "      <th>DH_Rate</th>\n",
       "      <th>TH_Rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>9.032219e-13</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47174.746248</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.695806</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002933</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.907041e-03</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47168.799678</td>\n",
       "      <td>0.064673</td>\n",
       "      <td>0.928514</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001389</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.04478</td>\n",
       "      <td>0.06523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>2.024492e-13</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>46351.638677</td>\n",
       "      <td>0.122135</td>\n",
       "      <td>0.995947</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001154</td>\n",
       "      <td>0.498937</td>\n",
       "      <td>0.551292</td>\n",
       "      <td>1.335249</td>\n",
       "      <td>0.379864</td>\n",
       "      <td>3.162586</td>\n",
       "      <td>0.068845</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>False</td>\n",
       "      <td>46352.004222</td>\n",
       "      <td>0.100309</td>\n",
       "      <td>0.612875</td>\n",
       "      <td>...</td>\n",
       "      <td>0.025081</td>\n",
       "      <td>0.493873</td>\n",
       "      <td>0.541742</td>\n",
       "      <td>1.322228</td>\n",
       "      <td>0.385386</td>\n",
       "      <td>3.05851</td>\n",
       "      <td>0.072872</td>\n",
       "      <td></td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.259925</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Gene      Method  Sequences  Codons  \\\n",
       "1  mammalian_REM2_codons.SA.FilterOutliers      BUSTED        175     627   \n",
       "2  mammalian_REM2_codons.SA.FilterOutliers   BUSTED+MH        175     627   \n",
       "3  mammalian_REM2_codons.SA.FilterOutliers     BUSTEDS        175     627   \n",
       "4  mammalian_REM2_codons.SA.FilterOutliers  BUSTEDS+MH        175     627   \n",
       "\n",
       "    LRT p-value  Bonferroni p-value Bonferroni significant          cAIC  \\\n",
       "1  9.032219e-13              0.0125                   True  47174.746248   \n",
       "2  1.907041e-03              0.0125                   True  47168.799678   \n",
       "3  2.024492e-13              0.0125                   True  46351.638677   \n",
       "4  5.000000e-01              0.0125                  False  46352.004222   \n",
       "\n",
       "         w1        p1  ...        p3      SRV1    SRV_p1      SRV2    SRV_p2  \\\n",
       "1  0.000000  0.695806  ...  0.002933                                           \n",
       "2  0.064673  0.928514  ...  0.001389                                           \n",
       "3  0.122135  0.995947  ...  0.001154  0.498937  0.551292  1.335249  0.379864   \n",
       "4  0.100309  0.612875  ...  0.025081  0.493873  0.541742  1.322228  0.385386   \n",
       "\n",
       "       SRV3    SRV_p3 Convergence_Issue   DH_Rate   TH_Rate  \n",
       "1                                                            \n",
       "2                                         0.04478   0.06523  \n",
       "3  3.162586  0.068845                                        \n",
       "4   3.05851  0.072872                    0.055794  0.259925  \n",
       "\n",
       "[4 rows x 23 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = pd.concat([df_BUSTEDS, df_BUSTEDS_MH, df_BUSTED, df_BUSTED_MH])\n",
    "result = result.fillna(\"\")\n",
    "result = result.sort_values(by=[\"Gene\", \"Method\"], ascending=True)\n",
    "result = result.reset_index(drop=True)\n",
    "result.index += 1\n",
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "comfortable-logic",
   "metadata": {},
   "source": [
    "## Calculate AIC weighted p-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "40999ca6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mammalian_REM2_codons.SA.FilterOutliers\n",
      "\tBest cAIC is: 46351.63867737226\n",
      "\tBest method: BUSTEDS\n",
      "p_ma is: 0.22721687386546213\n"
     ]
    }
   ],
   "source": [
    "df_main_holder = []\n",
    "for gene in set(result[\"Gene\"].tolist()):\n",
    "    print(gene)\n",
    "    df_holder = result[result[\"Gene\"] == gene]\n",
    "    min_cAIC = min(df_holder[\"cAIC\"].tolist())\n",
    "    print(\"\\tBest cAIC is:\", min_cAIC)\n",
    "    averaged = {}\n",
    "    for index, row in df_holder.iterrows():\n",
    "        method = row[\"Method\"]\n",
    "        cAIC   = row[\"cAIC\"]\n",
    "        pvalue = row[\"LRT p-value\"]\n",
    "        weight = math.exp (-0.5 * (cAIC - min_cAIC))\n",
    "        if cAIC == min_cAIC:\n",
    "            print(\"\\tBest method:\", method)\n",
    "        #end if\n",
    "        averaged[method] = {\n",
    "                            \"weight\": weight,\n",
    "                            \"original_pval\": pvalue\n",
    "        }\n",
    "    #end for\n",
    "    weight_sum = 0\n",
    "    for method in averaged.keys():\n",
    "        weight_sum += averaged[method][\"weight\"]\n",
    "    #end for\n",
    "    for method in averaged.keys():\n",
    "        averaged[method][\"normalized_weighted\"] = averaged[method][\"weight\"] / weight_sum\n",
    "    #end for\n",
    "    #calculate p_ma\n",
    "    p_ma = 0\n",
    "    for method in averaged.keys():\n",
    "        p_ma += averaged[method][\"original_pval\"] * averaged[method][\"normalized_weighted\"]\n",
    "    #end for\n",
    "    print(\"p_ma is:\", p_ma)\n",
    "    df_holder[\"p_value_averaged\"] = p_ma\n",
    "    df_main_holder.append(df_holder)\n",
    "#end for\n",
    "\n",
    "result = pd.concat(df_main_holder)\n",
    "result = result.fillna(\"\")\n",
    "result = result.sort_values(by=[\"Gene\", \"Method\"], ascending=True)\n",
    "result = result.reset_index(drop=True)\n",
    "result.index += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "focal-sullivan",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Method</th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Codons</th>\n",
       "      <th>LRT p-value</th>\n",
       "      <th>Bonferroni p-value</th>\n",
       "      <th>Bonferroni significant</th>\n",
       "      <th>cAIC</th>\n",
       "      <th>w1</th>\n",
       "      <th>p1</th>\n",
       "      <th>...</th>\n",
       "      <th>SRV1</th>\n",
       "      <th>SRV_p1</th>\n",
       "      <th>SRV2</th>\n",
       "      <th>SRV_p2</th>\n",
       "      <th>SRV3</th>\n",
       "      <th>SRV_p3</th>\n",
       "      <th>Convergence_Issue</th>\n",
       "      <th>DH_Rate</th>\n",
       "      <th>TH_Rate</th>\n",
       "      <th>p_value_averaged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>9.032219e-13</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47174.746248</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.695806</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.227217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.907041e-03</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47168.799678</td>\n",
       "      <td>0.064673</td>\n",
       "      <td>0.928514</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.04478</td>\n",
       "      <td>0.06523</td>\n",
       "      <td>0.227217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>2.024492e-13</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>46351.638677</td>\n",
       "      <td>0.122135</td>\n",
       "      <td>0.995947</td>\n",
       "      <td>...</td>\n",
       "      <td>0.498937</td>\n",
       "      <td>0.551292</td>\n",
       "      <td>1.335249</td>\n",
       "      <td>0.379864</td>\n",
       "      <td>3.162586</td>\n",
       "      <td>0.068845</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.227217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>False</td>\n",
       "      <td>46352.004222</td>\n",
       "      <td>0.100309</td>\n",
       "      <td>0.612875</td>\n",
       "      <td>...</td>\n",
       "      <td>0.493873</td>\n",
       "      <td>0.541742</td>\n",
       "      <td>1.322228</td>\n",
       "      <td>0.385386</td>\n",
       "      <td>3.05851</td>\n",
       "      <td>0.072872</td>\n",
       "      <td></td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.259925</td>\n",
       "      <td>0.227217</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Gene      Method  Sequences  Codons  \\\n",
       "1  mammalian_REM2_codons.SA.FilterOutliers      BUSTED        175     627   \n",
       "2  mammalian_REM2_codons.SA.FilterOutliers   BUSTED+MH        175     627   \n",
       "3  mammalian_REM2_codons.SA.FilterOutliers     BUSTEDS        175     627   \n",
       "4  mammalian_REM2_codons.SA.FilterOutliers  BUSTEDS+MH        175     627   \n",
       "\n",
       "    LRT p-value  Bonferroni p-value Bonferroni significant          cAIC  \\\n",
       "1  9.032219e-13              0.0125                   True  47174.746248   \n",
       "2  1.907041e-03              0.0125                   True  47168.799678   \n",
       "3  2.024492e-13              0.0125                   True  46351.638677   \n",
       "4  5.000000e-01              0.0125                  False  46352.004222   \n",
       "\n",
       "         w1        p1  ...      SRV1    SRV_p1      SRV2    SRV_p2      SRV3  \\\n",
       "1  0.000000  0.695806  ...                                                     \n",
       "2  0.064673  0.928514  ...                                                     \n",
       "3  0.122135  0.995947  ...  0.498937  0.551292  1.335249  0.379864  3.162586   \n",
       "4  0.100309  0.612875  ...  0.493873  0.541742  1.322228  0.385386   3.05851   \n",
       "\n",
       "     SRV_p3 Convergence_Issue   DH_Rate   TH_Rate p_value_averaged  \n",
       "1                                                         0.227217  \n",
       "2                               0.04478   0.06523         0.227217  \n",
       "3  0.068845                                               0.227217  \n",
       "4  0.072872                    0.055794  0.259925         0.227217  \n",
       "\n",
       "[4 rows x 24 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "15857068",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BUSTED': {'weight': 1.838493118342009e-179,\n",
       "  'original_pval': 9.032219416837961e-13,\n",
       "  'normalized_weighted': 1.0030198003967425e-179},\n",
       " 'BUSTED+MH': {'weight': 3.5953684039232464e-178,\n",
       "  'original_pval': 0.001907041306329127,\n",
       "  'normalized_weighted': 1.9615116656558484e-178},\n",
       " 'BUSTEDS': {'weight': 1.0,\n",
       "  'original_pval': 2.024491685403973e-13,\n",
       "  'normalized_weighted': 0.5455662522692967},\n",
       " 'BUSTEDS+MH': {'weight': 0.832957951193782,\n",
       "  'original_pval': 0.5,\n",
       "  'normalized_weighted': 0.4544337477307034}}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#for method in averaged.keys():\n",
    "#    averaged[method][\"normalized_weighted\"] = \n",
    "averaged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "amber-gregory",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/3l/76gtv9qd0v3_8hn3cb7hfbs40000gn/T/ipykernel_63887/2835796267.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_subset[\"Annotation\"][index] = \"Preferred\"\n"
     ]
    }
   ],
   "source": [
    "df_holder = []\n",
    "\n",
    "for g in set(result[\"Gene\"].to_list()):\n",
    "    #print(g)\n",
    "    df_subset = result[result[\"Gene\"] == g]\n",
    "    df_subset[\"Annotation\"] = \"\"\n",
    "    min_cAIC = min(df_subset[\"cAIC\"].to_list())\n",
    "    if len(df_subset[\"cAIC\"].to_list()) > 1:\n",
    "        second_smallest = sorted(df_subset[\"cAIC\"].to_list())[1]\n",
    "    else:\n",
    "        second_mallest = 0\n",
    "    #end if\n",
    "    for index, row in df_subset.iterrows():\n",
    "        caic = row[\"cAIC\"]\n",
    "        if caic == min_cAIC:\n",
    "            if second_smallest - caic > 5:\n",
    "                df_subset[\"Annotation\"][index] = \"Strongly Preferred\"\n",
    "            else:\n",
    "                df_subset[\"Annotation\"][index] = \"Preferred\"\n",
    "        #end if\n",
    "    #end for\n",
    "    df_holder.append(df_subset)\n",
    "#end for\n",
    "df_main = pd.concat(df_holder)\n",
    "df_main = df_main.reset_index(drop=True)\n",
    "df_main.index += 1\n",
    "#df_main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "brief-halloween",
   "metadata": {},
   "source": [
    "## Save table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "juvenile-depth",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving results to: mammalian_REM2_BUSTED_ModelTest.csv\n"
     ]
    }
   ],
   "source": [
    "#OUTPUT_CSV = \"Ebola.csv\"\n",
    "#OUTPUT_CSV = \"Zika.csv\"\n",
    "#OUTPUT_CSV = \"Monkeypox.csv\"\n",
    "#OUTPUT_CSV = TAG + \"-internal.csv\"\n",
    "\n",
    "print(\"Saving results to:\", OUTPUT_CSV)\n",
    "#result = result.T\n",
    "\n",
    "result.T.to_csv(os.path.join(\"..\",\n",
    "                           \"tables\",\n",
    "                           OUTPUT_CSV))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "premium-operation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Method</th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Codons</th>\n",
       "      <th>LRT p-value</th>\n",
       "      <th>Bonferroni p-value</th>\n",
       "      <th>Bonferroni significant</th>\n",
       "      <th>cAIC</th>\n",
       "      <th>w1</th>\n",
       "      <th>p1</th>\n",
       "      <th>...</th>\n",
       "      <th>SRV1</th>\n",
       "      <th>SRV_p1</th>\n",
       "      <th>SRV2</th>\n",
       "      <th>SRV_p2</th>\n",
       "      <th>SRV3</th>\n",
       "      <th>SRV_p3</th>\n",
       "      <th>Convergence_Issue</th>\n",
       "      <th>DH_Rate</th>\n",
       "      <th>TH_Rate</th>\n",
       "      <th>p_value_averaged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>9.032219e-13</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47174.746248</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.695806</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.227217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.907041e-03</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47168.799678</td>\n",
       "      <td>0.064673</td>\n",
       "      <td>0.928514</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.04478</td>\n",
       "      <td>0.06523</td>\n",
       "      <td>0.227217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>2.024492e-13</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>46351.638677</td>\n",
       "      <td>0.122135</td>\n",
       "      <td>0.995947</td>\n",
       "      <td>...</td>\n",
       "      <td>0.498937</td>\n",
       "      <td>0.551292</td>\n",
       "      <td>1.335249</td>\n",
       "      <td>0.379864</td>\n",
       "      <td>3.162586</td>\n",
       "      <td>0.068845</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.227217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>False</td>\n",
       "      <td>46352.004222</td>\n",
       "      <td>0.100309</td>\n",
       "      <td>0.612875</td>\n",
       "      <td>...</td>\n",
       "      <td>0.493873</td>\n",
       "      <td>0.541742</td>\n",
       "      <td>1.322228</td>\n",
       "      <td>0.385386</td>\n",
       "      <td>3.05851</td>\n",
       "      <td>0.072872</td>\n",
       "      <td></td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.259925</td>\n",
       "      <td>0.227217</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Gene      Method  Sequences  Codons  \\\n",
       "1  mammalian_REM2_codons.SA.FilterOutliers      BUSTED        175     627   \n",
       "2  mammalian_REM2_codons.SA.FilterOutliers   BUSTED+MH        175     627   \n",
       "3  mammalian_REM2_codons.SA.FilterOutliers     BUSTEDS        175     627   \n",
       "4  mammalian_REM2_codons.SA.FilterOutliers  BUSTEDS+MH        175     627   \n",
       "\n",
       "    LRT p-value  Bonferroni p-value Bonferroni significant          cAIC  \\\n",
       "1  9.032219e-13              0.0125                   True  47174.746248   \n",
       "2  1.907041e-03              0.0125                   True  47168.799678   \n",
       "3  2.024492e-13              0.0125                   True  46351.638677   \n",
       "4  5.000000e-01              0.0125                  False  46352.004222   \n",
       "\n",
       "         w1        p1  ...      SRV1    SRV_p1      SRV2    SRV_p2      SRV3  \\\n",
       "1  0.000000  0.695806  ...                                                     \n",
       "2  0.064673  0.928514  ...                                                     \n",
       "3  0.122135  0.995947  ...  0.498937  0.551292  1.335249  0.379864  3.162586   \n",
       "4  0.100309  0.612875  ...  0.493873  0.541742  1.322228  0.385386   3.05851   \n",
       "\n",
       "     SRV_p3 Convergence_Issue   DH_Rate   TH_Rate p_value_averaged  \n",
       "1                                                         0.227217  \n",
       "2                               0.04478   0.06523         0.227217  \n",
       "3  0.068845                                               0.227217  \n",
       "4  0.072872                    0.055794  0.259925         0.227217  \n",
       "\n",
       "[4 rows x 24 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0c0f1833",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = []\n",
    "for item in df_main[\"Gene\"].to_list():\n",
    "    gene_list.append(item.split(\"_\")[0])\n",
    "gene_set = set(gene_list)\n",
    "#gene_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4ea0114f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/3l/76gtv9qd0v3_8hn3cb7hfbs40000gn/T/ipykernel_63887/3009917599.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_main[\"Group\"][index] = str(gene.split(\"_\")[0])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Method</th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Codons</th>\n",
       "      <th>LRT p-value</th>\n",
       "      <th>Bonferroni p-value</th>\n",
       "      <th>Bonferroni significant</th>\n",
       "      <th>cAIC</th>\n",
       "      <th>w1</th>\n",
       "      <th>p1</th>\n",
       "      <th>...</th>\n",
       "      <th>SRV2</th>\n",
       "      <th>SRV_p2</th>\n",
       "      <th>SRV3</th>\n",
       "      <th>SRV_p3</th>\n",
       "      <th>Convergence_Issue</th>\n",
       "      <th>DH_Rate</th>\n",
       "      <th>TH_Rate</th>\n",
       "      <th>p_value_averaged</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>9.032219e-13</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47174.746248</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.695806</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.227217</td>\n",
       "      <td></td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.907041e-03</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47168.799678</td>\n",
       "      <td>0.064673</td>\n",
       "      <td>0.928514</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.04478</td>\n",
       "      <td>0.06523</td>\n",
       "      <td>0.227217</td>\n",
       "      <td></td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>2.024492e-13</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>46351.638677</td>\n",
       "      <td>0.122135</td>\n",
       "      <td>0.995947</td>\n",
       "      <td>...</td>\n",
       "      <td>1.335249</td>\n",
       "      <td>0.379864</td>\n",
       "      <td>3.162586</td>\n",
       "      <td>0.068845</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.227217</td>\n",
       "      <td>Preferred</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>False</td>\n",
       "      <td>46352.004222</td>\n",
       "      <td>0.100309</td>\n",
       "      <td>0.612875</td>\n",
       "      <td>...</td>\n",
       "      <td>1.322228</td>\n",
       "      <td>0.385386</td>\n",
       "      <td>3.05851</td>\n",
       "      <td>0.072872</td>\n",
       "      <td></td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.259925</td>\n",
       "      <td>0.227217</td>\n",
       "      <td></td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Gene      Method  Sequences  Codons  \\\n",
       "1  mammalian_REM2_codons.SA.FilterOutliers      BUSTED        175     627   \n",
       "2  mammalian_REM2_codons.SA.FilterOutliers   BUSTED+MH        175     627   \n",
       "3  mammalian_REM2_codons.SA.FilterOutliers     BUSTEDS        175     627   \n",
       "4  mammalian_REM2_codons.SA.FilterOutliers  BUSTEDS+MH        175     627   \n",
       "\n",
       "    LRT p-value  Bonferroni p-value Bonferroni significant          cAIC  \\\n",
       "1  9.032219e-13              0.0125                   True  47174.746248   \n",
       "2  1.907041e-03              0.0125                   True  47168.799678   \n",
       "3  2.024492e-13              0.0125                   True  46351.638677   \n",
       "4  5.000000e-01              0.0125                  False  46352.004222   \n",
       "\n",
       "         w1        p1  ...      SRV2    SRV_p2      SRV3    SRV_p3  \\\n",
       "1  0.000000  0.695806  ...                                           \n",
       "2  0.064673  0.928514  ...                                           \n",
       "3  0.122135  0.995947  ...  1.335249  0.379864  3.162586  0.068845   \n",
       "4  0.100309  0.612875  ...  1.322228  0.385386   3.05851  0.072872   \n",
       "\n",
       "  Convergence_Issue   DH_Rate   TH_Rate p_value_averaged Annotation      Group  \n",
       "1                                               0.227217             mammalian  \n",
       "2                     0.04478   0.06523         0.227217             mammalian  \n",
       "3                                               0.227217  Preferred  mammalian  \n",
       "4                    0.055794  0.259925         0.227217             mammalian  \n",
       "\n",
       "[4 rows x 26 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_main[\"Group\"] = \"\"\n",
    "for index, row in df_main.iterrows():\n",
    "    gene = row[\"Gene\"]\n",
    "    df_main[\"Group\"][index] = str(gene.split(\"_\")[0])\n",
    "df_main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e1fa996",
   "metadata": {},
   "source": [
    "## Visualize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "44cf86eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = result.rename(columns={\"LRT p-value\": \"p_value\"})\n",
    "df_main = df_main.rename(columns={\"LRT p-value\": \"p_value\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "014e93fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<style>\n",
       "  #altair-viz-cc7fdc4c4d2742bba0e8e0bc5189f8a1.vega-embed {\n",
       "    width: 100%;\n",
       "    display: flex;\n",
       "  }\n",
       "\n",
       "  #altair-viz-cc7fdc4c4d2742bba0e8e0bc5189f8a1.vega-embed details,\n",
       "  #altair-viz-cc7fdc4c4d2742bba0e8e0bc5189f8a1.vega-embed details summary {\n",
       "    position: relative;\n",
       "  }\n",
       "</style>\n",
       "<div id=\"altair-viz-cc7fdc4c4d2742bba0e8e0bc5189f8a1\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-cc7fdc4c4d2742bba0e8e0bc5189f8a1\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-cc7fdc4c4d2742bba0e8e0bc5189f8a1\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.16.3?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.16.3\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}, \"axis\": {\"labelFontSize\": 12, \"labelLimit\": 1000, \"titleFontSize\": 12}, \"scale\": {\"bandPaddingInner\": 0.05}}, \"layer\": [{\"mark\": {\"type\": \"rect\", \"opacity\": 0.9}, \"encoding\": {\"color\": {\"condition\": {\"test\": \"(datum.p_value <= 0.0125)\", \"value\": \"darkblue\"}, \"value\": \"lightgray\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}}, {\"mark\": {\"type\": \"text\"}, \"encoding\": {\"color\": {\"value\": \"white\"}, \"text\": {\"field\": \"Annotation\", \"type\": \"ordinal\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}}], \"data\": {\"name\": \"data-c2f6da0875eb2b9f5d999e48c60abb17\"}, \"height\": 1200, \"title\": \"mammalian_REM2-InternalBranches Genome Scan with BUSTED, LRT pValue threshold: 0.0125\", \"width\": 600, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.16.3.json\", \"datasets\": {\"data-c2f6da0875eb2b9f5d999e48c60abb17\": [{\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 9.032219416837961e-13, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47174.74624810679, \"w1\": 0.0, \"p1\": 0.6958061221543042, \"w2\": 0.4044151485405749, \"p2\": 0.3012613202589999, \"w3\": 23.15776038422468, \"p3\": 0.002932557586695876, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.001907041306329127, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47168.79967755279, \"w1\": 0.06467264227891008, \"p1\": 0.9285141565670676, \"w2\": 1.0, \"p2\": 0.07009664508720903, \"w3\": 35.2319122722087, \"p3\": 0.001389198345723368, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": 0.04478006132708046, \"TH_Rate\": 0.06522975636174812, \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 2.024491685403973e-13, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 46351.63867737226, \"w1\": 0.1221351888166984, \"p1\": 0.9959468369082943, \"w2\": 1.0, \"p2\": 0.002898972414528135, \"w3\": 50.74823182825114, \"p3\": 0.001154190677177535, \"SRV1\": 0.4989374204609935, \"SRV_p1\": 0.5512916806271009, \"SRV2\": 1.335249472783704, \"SRV_p2\": 0.3798636395406089, \"SRV3\": 3.162586085815315, \"SRV_p3\": 0.06884467983229028, \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"Preferred\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.5, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 46352.00422160596, \"w1\": 0.100308808460714, \"p1\": 0.6128754322373415, \"w2\": 0.1240906545234857, \"p2\": 0.3620440607754455, \"w3\": 1.0, \"p3\": 0.02508050698721298, \"SRV1\": 0.4938725957944278, \"SRV_p1\": 0.5417419720260168, \"SRV2\": 1.322227742200831, \"SRV_p2\": 0.3853856472019641, \"SRV3\": 3.058509815200658, \"SRV_p3\": 0.07287238077201906, \"Convergence_Issue\": \"\", \"DH_Rate\": 0.0557943840619701, \"TH_Rate\": 0.2599252630852945, \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"\", \"Group\": \"mammalian\"}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source = df_main\n",
    "pval_threshold = 0.05 / Tests\n",
    "\n",
    "heatmap = alt.Chart(source).mark_rect(opacity=0.9).encode(\n",
    "    y = alt.Y('Gene'),\n",
    "    x = alt.X('Method', axis=alt.Axis(labelAngle=-60)),\n",
    "    color = alt.condition((alt.datum.p_value <= pval_threshold), \n",
    "                                alt.ColorValue('darkblue'), \n",
    "                                alt.ColorValue('lightgray')) \n",
    ").properties(\n",
    "    width=600,\n",
    "    height=1200\n",
    ")\n",
    "\n",
    "text = heatmap.mark_text().encode(\n",
    "    text='Annotation:O',\n",
    "    color=alt.value('white'))\n",
    "\n",
    "chart = alt.layer(heatmap, text).configure_axis(\n",
    "    labelFontSize=12,\n",
    "    titleFontSize=12,\n",
    "    labelLimit = 1000\n",
    ").configure_scale(bandPaddingInner=0.05).properties(\n",
    "    title=TAG+\"-InternalBranches Genome Scan with BUSTED, LRT pValue threshold: \" + str(pval_threshold))\n",
    "\n",
    "chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bee9914f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<style>\n",
       "  #altair-viz-12b741c3aa4b4028960ad89f4f8d74bc.vega-embed {\n",
       "    width: 100%;\n",
       "    display: flex;\n",
       "  }\n",
       "\n",
       "  #altair-viz-12b741c3aa4b4028960ad89f4f8d74bc.vega-embed details,\n",
       "  #altair-viz-12b741c3aa4b4028960ad89f4f8d74bc.vega-embed details summary {\n",
       "    position: relative;\n",
       "  }\n",
       "</style>\n",
       "<div id=\"altair-viz-12b741c3aa4b4028960ad89f4f8d74bc\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-12b741c3aa4b4028960ad89f4f8d74bc\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-12b741c3aa4b4028960ad89f4f8d74bc\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.16.3?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.16.3\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}, \"axis\": {\"labelFontSize\": 12, \"labelLimit\": 1000, \"titleFontSize\": 12}, \"scale\": {\"bandPaddingInner\": 0.05}}, \"layer\": [{\"mark\": {\"type\": \"rect\", \"opacity\": 0.75}, \"encoding\": {\"color\": {\"condition\": {\"test\": \"(datum.p_value_averaged <= 0.05)\", \"value\": \"darkblue\"}, \"value\": \"gray\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}}, {\"mark\": {\"type\": \"text\"}, \"encoding\": {\"color\": {\"value\": \"white\"}, \"text\": {\"field\": \"Annotation\", \"type\": \"ordinal\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}}], \"data\": {\"name\": \"data-c2f6da0875eb2b9f5d999e48c60abb17\"}, \"height\": 1200, \"title\": \"mammalian_REM2 on Internal Branches Genome Scan with BUSTED, LRT p-value threshold: 0.05 on the AIC-weighted pvalue \", \"width\": 800, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.16.3.json\", \"datasets\": {\"data-c2f6da0875eb2b9f5d999e48c60abb17\": [{\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 9.032219416837961e-13, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47174.74624810679, \"w1\": 0.0, \"p1\": 0.6958061221543042, \"w2\": 0.4044151485405749, \"p2\": 0.3012613202589999, \"w3\": 23.15776038422468, \"p3\": 0.002932557586695876, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.001907041306329127, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47168.79967755279, \"w1\": 0.06467264227891008, \"p1\": 0.9285141565670676, \"w2\": 1.0, \"p2\": 0.07009664508720903, \"w3\": 35.2319122722087, \"p3\": 0.001389198345723368, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": 0.04478006132708046, \"TH_Rate\": 0.06522975636174812, \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 2.024491685403973e-13, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 46351.63867737226, \"w1\": 0.1221351888166984, \"p1\": 0.9959468369082943, \"w2\": 1.0, \"p2\": 0.002898972414528135, \"w3\": 50.74823182825114, \"p3\": 0.001154190677177535, \"SRV1\": 0.4989374204609935, \"SRV_p1\": 0.5512916806271009, \"SRV2\": 1.335249472783704, \"SRV_p2\": 0.3798636395406089, \"SRV3\": 3.162586085815315, \"SRV_p3\": 0.06884467983229028, \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"Preferred\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.5, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 46352.00422160596, \"w1\": 0.100308808460714, \"p1\": 0.6128754322373415, \"w2\": 0.1240906545234857, \"p2\": 0.3620440607754455, \"w3\": 1.0, \"p3\": 0.02508050698721298, \"SRV1\": 0.4938725957944278, \"SRV_p1\": 0.5417419720260168, \"SRV2\": 1.322227742200831, \"SRV_p2\": 0.3853856472019641, \"SRV3\": 3.058509815200658, \"SRV_p3\": 0.07287238077201906, \"Convergence_Issue\": \"\", \"DH_Rate\": 0.0557943840619701, \"TH_Rate\": 0.2599252630852945, \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"\", \"Group\": \"mammalian\"}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source = df_main\n",
    "\n",
    "pval_threshold = 0.05\n",
    "\n",
    "heatmap = alt.Chart(source).mark_rect(opacity=0.75).encode(\n",
    "    y = alt.Y('Gene'),\n",
    "    x = alt.X('Method', axis=alt.Axis(labelAngle=-60)),\n",
    "    color = alt.condition((alt.datum.p_value_averaged <= pval_threshold), \n",
    "                                alt.ColorValue('darkblue'), \n",
    "                                alt.ColorValue('gray')) \n",
    ").properties(\n",
    "    width=800,\n",
    "    height=1200\n",
    ")\n",
    "\n",
    "text = heatmap.mark_text().encode(\n",
    "    text='Annotation:O',\n",
    "    color=alt.value('white'))\n",
    "\n",
    "chart = alt.layer(heatmap, text).configure_axis(\n",
    "    labelFontSize=12,\n",
    "    titleFontSize=12,\n",
    "    labelLimit = 1000\n",
    ").configure_scale(bandPaddingInner=0.05).properties(\n",
    "    title=TAG+\" on Internal Branches Genome Scan with BUSTED, LRT p-value threshold: \" + str(pval_threshold) + \" on the AIC-weighted pvalue \")\n",
    "\n",
    "chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0e045c1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<style>\n",
       "  #altair-viz-ea8dffa0e44c49cdbe74e281502ea960.vega-embed {\n",
       "    width: 100%;\n",
       "    display: flex;\n",
       "  }\n",
       "\n",
       "  #altair-viz-ea8dffa0e44c49cdbe74e281502ea960.vega-embed details,\n",
       "  #altair-viz-ea8dffa0e44c49cdbe74e281502ea960.vega-embed details summary {\n",
       "    position: relative;\n",
       "  }\n",
       "</style>\n",
       "<div id=\"altair-viz-ea8dffa0e44c49cdbe74e281502ea960\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-ea8dffa0e44c49cdbe74e281502ea960\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-ea8dffa0e44c49cdbe74e281502ea960\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.16.3?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.16.3\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}, \"axis\": {\"labelFontSize\": 12, \"labelLimit\": 1000, \"titleFontSize\": 12}, \"scale\": {\"bandPaddingInner\": 0.05}}, \"layer\": [{\"mark\": {\"type\": \"rect\", \"opacity\": 0.75}, \"encoding\": {\"color\": {\"condition\": {\"test\": \"(datum.p_value_averaged <= 0.05)\", \"value\": \"darkblue\"}, \"value\": \"orange\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}}, {\"mark\": {\"type\": \"text\"}, \"encoding\": {\"color\": {\"value\": \"white\"}, \"text\": {\"field\": \"Annotation\", \"type\": \"ordinal\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}}], \"data\": {\"name\": \"data-c2f6da0875eb2b9f5d999e48c60abb17\"}, \"encoding\": {}, \"height\": 1200, \"title\": \"mammalian_REM2 on Internal Branches Genome Scan with BUSTED, LRT p-value threshold: 0.05 on the AIC-weighted pvalue \", \"width\": 800, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.16.3.json\", \"datasets\": {\"data-c2f6da0875eb2b9f5d999e48c60abb17\": [{\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 9.032219416837961e-13, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47174.74624810679, \"w1\": 0.0, \"p1\": 0.6958061221543042, \"w2\": 0.4044151485405749, \"p2\": 0.3012613202589999, \"w3\": 23.15776038422468, \"p3\": 0.002932557586695876, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.001907041306329127, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47168.79967755279, \"w1\": 0.06467264227891008, \"p1\": 0.9285141565670676, \"w2\": 1.0, \"p2\": 0.07009664508720903, \"w3\": 35.2319122722087, \"p3\": 0.001389198345723368, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": 0.04478006132708046, \"TH_Rate\": 0.06522975636174812, \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 2.024491685403973e-13, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 46351.63867737226, \"w1\": 0.1221351888166984, \"p1\": 0.9959468369082943, \"w2\": 1.0, \"p2\": 0.002898972414528135, \"w3\": 50.74823182825114, \"p3\": 0.001154190677177535, \"SRV1\": 0.4989374204609935, \"SRV_p1\": 0.5512916806271009, \"SRV2\": 1.335249472783704, \"SRV_p2\": 0.3798636395406089, \"SRV3\": 3.162586085815315, \"SRV_p3\": 0.06884467983229028, \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"Preferred\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.5, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 46352.00422160596, \"w1\": 0.100308808460714, \"p1\": 0.6128754322373415, \"w2\": 0.1240906545234857, \"p2\": 0.3620440607754455, \"w3\": 1.0, \"p3\": 0.02508050698721298, \"SRV1\": 0.4938725957944278, \"SRV_p1\": 0.5417419720260168, \"SRV2\": 1.322227742200831, \"SRV_p2\": 0.3853856472019641, \"SRV3\": 3.058509815200658, \"SRV_p3\": 0.07287238077201906, \"Convergence_Issue\": \"\", \"DH_Rate\": 0.0557943840619701, \"TH_Rate\": 0.2599252630852945, \"p_value_averaged\": 0.22721687386546213, \"Annotation\": \"\", \"Group\": \"mammalian\"}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source = df_main.copy()\n",
    "\n",
    "pval_threshold = 0.05\n",
    "\n",
    "heatmap = alt.Chart(source).mark_rect(opacity=0.75).encode(\n",
    "    y = alt.Y('Gene'),\n",
    "    x = alt.X('Method', axis=alt.Axis(labelAngle=-60)),\n",
    "    color = alt.condition((alt.datum.p_value_averaged <= pval_threshold), \n",
    "                                alt.ColorValue('darkblue'), \n",
    "                                alt.ColorValue('orange')),\n",
    ").properties(\n",
    "    width=800,\n",
    "    height=1200\n",
    ")\n",
    "\n",
    "text = heatmap.mark_text().encode(\n",
    "    text='Annotation:O',\n",
    "    color=alt.value('white'))\n",
    "\n",
    "chart = alt.layer(heatmap, text).encode(\n",
    ").configure_axis(\n",
    "    labelFontSize=12,\n",
    "    titleFontSize=12,\n",
    "    labelLimit = 1000,\n",
    ").configure_scale(bandPaddingInner=0.05).properties(\n",
    "    title=TAG+\" on Internal Branches Genome Scan with BUSTED, LRT p-value threshold: \" + str(pval_threshold) + \" on the AIC-weighted pvalue \")\n",
    "\n",
    "chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "058d1746-fba2-47b5-ba91-017e6861578d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
