{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "complex-berkeley",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "radical-anxiety",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import csv\n",
    "import os\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "import math\n",
    "import altair as alt\n",
    "import glob\n",
    "import sys\n",
    "\n",
    "module_path = os.path.abspath(os.path.join('..', 'scripts'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "#end if\n",
    "\n",
    "import utils\n",
    "from utils import read_json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "decreased-tract",
   "metadata": {},
   "source": [
    "## Declares"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "outer-supplier",
   "metadata": {},
   "outputs": [],
   "source": [
    "TAG = \"mammalian_REM2\"\n",
    "DATA_DIRECTORY = os.path.join(\"..\", \"results\", TAG)\n",
    "\n",
    "JSONS_BS = glob.glob(os.path.join(DATA_DIRECTORY, \"*.BUSTEDS.json\"))\n",
    "JSONS_BSMH = glob.glob(os.path.join(DATA_DIRECTORY, \"*.BUSTEDS+MH.json\"))\n",
    "JSONS_BASE = glob.glob(os.path.join(DATA_DIRECTORY, \"*.BUSTED.json\"))\n",
    "JSONS_BMH = glob.glob(os.path.join(DATA_DIRECTORY, \"*.BUSTED+MH.json\"))\n",
    "\n",
    "# Create tables folder\n",
    "OUTPUT_CSV = TAG + \"_BUSTED_ModelTest.csv\"\n",
    "\n",
    "ER_Threshold = 10\n",
    "ER_Threshold_loose = 1\n",
    "\n",
    "pval_Threshold = 0.05\n",
    "\n",
    "Tests = 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "sealed-flexibility",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define function to calculate cv\n",
    "#cv = lambda x: np.std(x, ddof=1) / np.mean(x) * 100 \n",
    "cv = lambda x: np.std(x) / np.mean(x)\n",
    "pctchg = lambda a, b: (a / b) * 100\n",
    "\n",
    "\"\"\"\n",
    "def read_json(filename):\n",
    "    if os.stat(filename).st_size == 0: \n",
    "        print(\"# -- Error -- file is empty:\", filename)\n",
    "        return []\n",
    "    #end if\n",
    "    with open(filename, \"r\") as fh:\n",
    "        json_data = json.load(fh)\n",
    "    fh.close()\n",
    "    return json_data\n",
    "#end method\n",
    "\"\"\"\n",
    "\n",
    "def convergence_issue(json_data):\n",
    "    # For example if logL for a more complex model is LOWER than it is for a less complex model.\n",
    "    convergence_issue = False \n",
    "    return \"\"\n",
    "    \"\"\"\n",
    "    Constrained model\n",
    "    MG94xREV with separate rates for branch sets\n",
    "    Nucleotide GTR\n",
    "    Unconstrained model\n",
    "    \"\"\"\n",
    "    \n",
    "    _models = [\"Nucleotide GTR\", \n",
    "                    \"MG94xREV with separate rates for branch sets\", \n",
    "                    \"Constrained model\", \n",
    "                    \"Unconstrained model\"]\n",
    "    model_LL = {}\n",
    "    \n",
    "    for _model in _models:\n",
    "        print(_model)\n",
    "        _LL = json_data[\"fits\"][_model][\"Log Likelihood\"]\n",
    "\n",
    "        \n",
    "    \"\"\"\n",
    "    return\n",
    "    #Nucleotide_GTR = BS_data[\"fits\"][\"Nucleotide GTR\"].get(\"Log Likelihood\", 10)\n",
    "    #MG94_REV = BS_data[\"fits\"][\"MG94xREV with separate rates for branch sets\"].get(\"Log Likelihood\",10)\n",
    "    try:\n",
    "        Constrained_model = BS_data[\"fits\"][\"Constrained model\"].get(\"Log Likelihood\", 10)\n",
    "    except:\n",
    "        Constrained_model = 10\n",
    "    #end try\n",
    "    Unconstrained_model = BS_data[\"fits\"][\"Unconstrained model\"].get(\"Log Likelihood\", 10)\n",
    "    if Nucleotide_GTR < MG94_REV and MG94_REV < Constrained_model and Constrained_model < Unconstrained_model:\n",
    "        pass\n",
    "    else:\n",
    "        issue = True\n",
    "    #end if\n",
    "    if issue == False:\n",
    "        return \"\"\n",
    "    else:\n",
    "        return \"ISSUE DETECTED\"\n",
    "    #end if\n",
    "    \"\"\"\n",
    "#end method\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def convergence_issue_old(BS_data):\n",
    "    # For example if logL for a more complex model is LOWER than it is for a less complex model.\n",
    "    convergence_issue = False \n",
    "    \n",
    "    #Nucleotide_GTR = BS_data[\"fits\"][\"Nucleotide GTR\"].get(\"Log Likelihood\", 10)\n",
    "    #MG94_REV = BS_data[\"fits\"][\"MG94xREV with separate rates for branch sets\"].get(\"Log Likelihood\",10)\n",
    "    \n",
    "    try:\n",
    "        Constrained_model = BS_data[\"fits\"][\"Constrained model\"].get(\"Log Likelihood\", 10)\n",
    "    except:\n",
    "        Constrained_model = 10\n",
    "    #end try\n",
    "    \n",
    "    Unconstrained_model = BS_data[\"fits\"][\"Unconstrained model\"].get(\"Log Likelihood\", 10)\n",
    "    \n",
    "    if Nucleotide_GTR < MG94_REV and MG94_REV < Constrained_model and Constrained_model < Unconstrained_model:\n",
    "        pass\n",
    "    else:\n",
    "        issue = True\n",
    "    #end if\n",
    "    \n",
    "    if issue == False:\n",
    "        return \"\"\n",
    "    else:\n",
    "        return \"ISSUE DETECTED\"\n",
    "    #end if\n",
    "#end method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "similar-active",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process(FILES, fileending, method, pval_Threshold, Tests):\n",
    "    df_dict = {}\n",
    "    Bonferroni_pval = pval_Threshold / Tests\n",
    "    for item in FILES:\n",
    "        print(\"\\t Processing:\", item)\n",
    "        basename = \"\"\n",
    "        if fileending in os.path.basename(item):\n",
    "             basename = os.path.basename(item).replace(fileending, \"\").replace(\".best-gard\", \"\")\n",
    "             #basename = os.path.basename(item)\n",
    "             #print(\"\\t Basename:\", basename)\n",
    "        #end if\n",
    "        for fext in [\".phy\", \".fasta\", \".nex\", \"-align-dna.fas\", \"-Aligned-DNA.fas\"]:\n",
    "            basename = basename.replace(fext, \"\")\n",
    "        #end for\n",
    "        json_data = read_json(item) # Read json\n",
    "        if json_data == []:\n",
    "            continue # Empty file\n",
    "        #end if\n",
    "        df_dict[basename] = {\"Method\": method}\n",
    "        df_dict[basename].update({\"Sequences\": json_data[\"input\"][\"number of sequences\"]})\n",
    "        df_dict[basename].update({\"Codons\": json_data[\"input\"][\"number of sites\"]})\n",
    "        df_dict[basename].update({\"LRT p-value\": json_data[\"test results\"][\"p-value\"]})\n",
    "        df_dict[basename].update({\"Bonferroni p-value\": Bonferroni_pval})\n",
    "        Bonferroni_sig = False\n",
    "        if json_data[\"test results\"][\"p-value\"] <= Bonferroni_pval:\n",
    "            Bonferroni_sig = True\n",
    "        #end if\n",
    "        # Bonferroni Test\n",
    "        df_dict[basename].update({\"Bonferroni significant\": str(Bonferroni_sig)})\n",
    "        # cAIC\n",
    "        df_dict[basename].update({\"cAIC\": json_data[\"fits\"][\"Unconstrained model\"][\"AIC-c\"]})\n",
    "        #Omegas and proportions\n",
    "        data = json_data[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Test\"]\n",
    "        w1 = data[\"0\"][\"omega\"]\n",
    "        p1 = data[\"0\"][\"proportion\"]\n",
    "        w2 = data[\"1\"][\"omega\"]\n",
    "        p2 = data[\"1\"][\"proportion\"]\n",
    "        w3 = data[\"2\"][\"omega\"]\n",
    "        p3 = data[\"2\"][\"proportion\"]\n",
    "        df_dict[basename].update({\"w1\": w1, \"p1\": p1})\n",
    "        df_dict[basename].update({\"w2\": w2, \"p2\": p2})\n",
    "        df_dict[basename].update({\"w3\": w3, \"p3\": p3})\n",
    "        # SRV rates and proportions\n",
    "        if method == \"BUSTEDS+MH\" or method == \"BUSTEDS\":\n",
    "            data = json_data[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Synonymous site-to-site rates\"]\n",
    "            s1 = data[\"0\"][\"rate\"]\n",
    "            s_p1 = data[\"0\"][\"proportion\"]\n",
    "            s2 = data[\"1\"][\"rate\"]\n",
    "            s_p2 = data[\"1\"][\"proportion\"]\n",
    "            s3 = data[\"2\"][\"rate\"]\n",
    "            s_p3 = data[\"2\"][\"proportion\"]\n",
    "            df_dict[basename].update({\"SRV1\": s1, \"SRV_p1\": s_p1})\n",
    "            df_dict[basename].update({\"SRV2\": s2, \"SRV_p2\": s_p2})\n",
    "            df_dict[basename].update({\"SRV3\": s3, \"SRV_p3\": s_p3})\n",
    "        #end if\n",
    "        # DH rate, TH rate, TH_SI rate\n",
    "        if method == \"BUSTEDS+MH\" or method == \"BUSTED+MH\":\n",
    "            df_dict[basename].update({\"DH_Rate\": float(json_data[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"rate at which 2 nucleotides are changed instantly within a single codon\"])})\n",
    "            df_dict[basename].update({\"TH_Rate\": float(json_data[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"rate at which 3 nucleotides are changed instantly within a single codon\"])})\n",
    "            #df_dict[basename].update({\"TH_Rate_SI\": float(json_data[\"fits\"][\"Unconstrained model\"][\"rate at which 3 nucleotides are changed instantly within a single codon between synonymous codon islands\"])})\n",
    "        #end if\n",
    "        # Convergence issues check\n",
    "        df_dict[basename].update({\"Convergence_Issue\": convergence_issue(json_data)})\n",
    "    # end for\n",
    "    df = pd.DataFrame.from_dict(df_dict, orient=\"index\")\n",
    "    df = df.reset_index()\n",
    "    df.index += 1\n",
    "    df.rename(columns={'index': 'Gene'}, inplace = True)\n",
    "    return df\n",
    "#end method\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "defa9746-0504-42cb-9b81-ca70e1ae9ad6",
   "metadata": {},
   "source": [
    "## Process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "academic-shoot",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Processing BUSTED[S] files\n",
      "\t Processing: ..\\results\\mammalian_REM2\\mammalian_REM2_codons.SA.FilterOutliers.fasta.BUSTEDS.json\n",
      "\n",
      "# Processing BUSTED[S]-MH files\n",
      "\t Processing: ..\\results\\mammalian_REM2\\mammalian_REM2_codons.SA.FilterOutliers.fasta.BUSTEDS+MH.json\n",
      "\n",
      "# Processing BUSTED files\n",
      "\t Processing: ..\\results\\mammalian_REM2\\mammalian_REM2_codons.SA.FilterOutliers.fasta.BUSTED.json\n",
      "\n",
      "# Processing BUSTED-MH files\n",
      "\t Processing: ..\\results\\mammalian_REM2\\mammalian_REM2_codons.SA.FilterOutliers.fasta.BUSTED+MH.json\n"
     ]
    }
   ],
   "source": [
    "print(\"# Processing BUSTED[S] files\")\n",
    "df_BUSTEDS = process(JSONS_BS , \".BUSTEDS.json\", \"BUSTEDS\", pval_Threshold, Tests)\n",
    "\n",
    "print(\"\\n# Processing BUSTED[S]-MH files\")\n",
    "df_BUSTEDS_MH = process(JSONS_BSMH, \".BUSTEDS+MH.json\", \"BUSTEDS+MH\", pval_Threshold, Tests)\n",
    "\n",
    "print(\"\\n# Processing BUSTED files\")\n",
    "df_BUSTED = process(JSONS_BASE , \".BUSTED.json\", \"BUSTED\", pval_Threshold, Tests)\n",
    "print(\"\\n# Processing BUSTED-MH files\")\n",
    "df_BUSTED_MH = process(JSONS_BMH, \".BUSTED+MH.json\", \"BUSTED+MH\", pval_Threshold, Tests)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5509d43",
   "metadata": {},
   "source": [
    "## Concatenate tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "908b0296",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Method</th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Codons</th>\n",
       "      <th>LRT p-value</th>\n",
       "      <th>Bonferroni p-value</th>\n",
       "      <th>Bonferroni significant</th>\n",
       "      <th>cAIC</th>\n",
       "      <th>w1</th>\n",
       "      <th>p1</th>\n",
       "      <th>...</th>\n",
       "      <th>p3</th>\n",
       "      <th>SRV1</th>\n",
       "      <th>SRV_p1</th>\n",
       "      <th>SRV2</th>\n",
       "      <th>SRV_p2</th>\n",
       "      <th>SRV3</th>\n",
       "      <th>SRV_p3</th>\n",
       "      <th>Convergence_Issue</th>\n",
       "      <th>DH_Rate</th>\n",
       "      <th>TH_Rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>3.268630e-11</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47181.924003</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.241485</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004426</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.935679e-03</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47168.822954</td>\n",
       "      <td>0.064925</td>\n",
       "      <td>0.928944</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001418</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.044895</td>\n",
       "      <td>0.067012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>9.103829e-15</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>46345.197711</td>\n",
       "      <td>0.077807</td>\n",
       "      <td>0.891885</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000565</td>\n",
       "      <td>0.497895</td>\n",
       "      <td>0.552027</td>\n",
       "      <td>1.340823</td>\n",
       "      <td>0.37766</td>\n",
       "      <td>3.111401</td>\n",
       "      <td>0.070314</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.262289e-02</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>False</td>\n",
       "      <td>46344.421537</td>\n",
       "      <td>0.120585</td>\n",
       "      <td>0.993374</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000698</td>\n",
       "      <td>0.499453</td>\n",
       "      <td>0.55085</td>\n",
       "      <td>1.335909</td>\n",
       "      <td>0.379458</td>\n",
       "      <td>3.127391</td>\n",
       "      <td>0.069692</td>\n",
       "      <td></td>\n",
       "      <td>0.032234</td>\n",
       "      <td>0.14427</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Gene      Method  Sequences  Codons  \\\n",
       "1  mammalian_REM2_codons.SA.FilterOutliers      BUSTED        175     627   \n",
       "2  mammalian_REM2_codons.SA.FilterOutliers   BUSTED+MH        175     627   \n",
       "3  mammalian_REM2_codons.SA.FilterOutliers     BUSTEDS        175     627   \n",
       "4  mammalian_REM2_codons.SA.FilterOutliers  BUSTEDS+MH        175     627   \n",
       "\n",
       "    LRT p-value  Bonferroni p-value Bonferroni significant          cAIC  \\\n",
       "1  3.268630e-11              0.0125                   True  47181.924003   \n",
       "2  1.935679e-03              0.0125                   True  47168.822954   \n",
       "3  9.103829e-15              0.0125                   True  46345.197711   \n",
       "4  1.262289e-02              0.0125                  False  46344.421537   \n",
       "\n",
       "         w1        p1  ...        p3      SRV1    SRV_p1      SRV2    SRV_p2  \\\n",
       "1  0.000000  0.241485  ...  0.004426                                           \n",
       "2  0.064925  0.928944  ...  0.001418                                           \n",
       "3  0.077807  0.891885  ...  0.000565  0.497895  0.552027  1.340823   0.37766   \n",
       "4  0.120585  0.993374  ...  0.000698  0.499453   0.55085  1.335909  0.379458   \n",
       "\n",
       "       SRV3    SRV_p3 Convergence_Issue   DH_Rate   TH_Rate  \n",
       "1                                                            \n",
       "2                                        0.044895  0.067012  \n",
       "3  3.111401  0.070314                                        \n",
       "4  3.127391  0.069692                    0.032234   0.14427  \n",
       "\n",
       "[4 rows x 23 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = pd.concat([df_BUSTEDS, df_BUSTEDS_MH, df_BUSTED, df_BUSTED_MH])\n",
    "result = result.fillna(\"\")\n",
    "result = result.sort_values(by=[\"Gene\", \"Method\"], ascending=True)\n",
    "result = result.reset_index(drop=True)\n",
    "result.index += 1\n",
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "comfortable-logic",
   "metadata": {},
   "source": [
    "## Calculate AIC weighted p-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "40999ca6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mammalian_REM2_codons.SA.FilterOutliers\n",
      "\tBest cAIC is: 46344.42153659258\n",
      "\tBest method: BUSTEDS+MH\n",
      "p_ma is: 0.007520994999594324\n"
     ]
    }
   ],
   "source": [
    "df_main_holder = []\n",
    "for gene in set(result[\"Gene\"].tolist()):\n",
    "    print(gene)\n",
    "    df_holder = result[result[\"Gene\"] == gene]\n",
    "    min_cAIC = min(df_holder[\"cAIC\"].tolist())\n",
    "    print(\"\\tBest cAIC is:\", min_cAIC)\n",
    "    averaged = {}\n",
    "    for index, row in df_holder.iterrows():\n",
    "        method = row[\"Method\"]\n",
    "        cAIC   = row[\"cAIC\"]\n",
    "        pvalue = row[\"LRT p-value\"]\n",
    "        weight = math.exp (-0.5 * (cAIC - min_cAIC))\n",
    "        if cAIC == min_cAIC:\n",
    "            print(\"\\tBest method:\", method)\n",
    "        #end if\n",
    "        averaged[method] = {\n",
    "                            \"weight\": weight,\n",
    "                            \"original_pval\": pvalue\n",
    "        }\n",
    "    #end for\n",
    "    weight_sum = 0\n",
    "    for method in averaged.keys():\n",
    "        weight_sum += averaged[method][\"weight\"]\n",
    "    #end for\n",
    "    for method in averaged.keys():\n",
    "        averaged[method][\"normalized_weighted\"] = averaged[method][\"weight\"] / weight_sum\n",
    "    #end for\n",
    "    #calculate p_ma\n",
    "    p_ma = 0\n",
    "    for method in averaged.keys():\n",
    "        p_ma += averaged[method][\"original_pval\"] * averaged[method][\"normalized_weighted\"]\n",
    "    #end for\n",
    "    print(\"p_ma is:\", p_ma)\n",
    "    df_holder[\"p_value_averaged\"] = p_ma\n",
    "    df_main_holder.append(df_holder)\n",
    "#end for\n",
    "\n",
    "result = pd.concat(df_main_holder)\n",
    "result = result.fillna(\"\")\n",
    "result = result.sort_values(by=[\"Gene\", \"Method\"], ascending=True)\n",
    "result = result.reset_index(drop=True)\n",
    "result.index += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "focal-sullivan",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Method</th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Codons</th>\n",
       "      <th>LRT p-value</th>\n",
       "      <th>Bonferroni p-value</th>\n",
       "      <th>Bonferroni significant</th>\n",
       "      <th>cAIC</th>\n",
       "      <th>w1</th>\n",
       "      <th>p1</th>\n",
       "      <th>...</th>\n",
       "      <th>SRV1</th>\n",
       "      <th>SRV_p1</th>\n",
       "      <th>SRV2</th>\n",
       "      <th>SRV_p2</th>\n",
       "      <th>SRV3</th>\n",
       "      <th>SRV_p3</th>\n",
       "      <th>Convergence_Issue</th>\n",
       "      <th>DH_Rate</th>\n",
       "      <th>TH_Rate</th>\n",
       "      <th>p_value_averaged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>3.268630e-11</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47181.924003</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.241485</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.007521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.935679e-03</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47168.822954</td>\n",
       "      <td>0.064925</td>\n",
       "      <td>0.928944</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.044895</td>\n",
       "      <td>0.067012</td>\n",
       "      <td>0.007521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>9.103829e-15</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>46345.197711</td>\n",
       "      <td>0.077807</td>\n",
       "      <td>0.891885</td>\n",
       "      <td>...</td>\n",
       "      <td>0.497895</td>\n",
       "      <td>0.552027</td>\n",
       "      <td>1.340823</td>\n",
       "      <td>0.37766</td>\n",
       "      <td>3.111401</td>\n",
       "      <td>0.070314</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.007521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.262289e-02</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>False</td>\n",
       "      <td>46344.421537</td>\n",
       "      <td>0.120585</td>\n",
       "      <td>0.993374</td>\n",
       "      <td>...</td>\n",
       "      <td>0.499453</td>\n",
       "      <td>0.55085</td>\n",
       "      <td>1.335909</td>\n",
       "      <td>0.379458</td>\n",
       "      <td>3.127391</td>\n",
       "      <td>0.069692</td>\n",
       "      <td></td>\n",
       "      <td>0.032234</td>\n",
       "      <td>0.14427</td>\n",
       "      <td>0.007521</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Gene      Method  Sequences  Codons  \\\n",
       "1  mammalian_REM2_codons.SA.FilterOutliers      BUSTED        175     627   \n",
       "2  mammalian_REM2_codons.SA.FilterOutliers   BUSTED+MH        175     627   \n",
       "3  mammalian_REM2_codons.SA.FilterOutliers     BUSTEDS        175     627   \n",
       "4  mammalian_REM2_codons.SA.FilterOutliers  BUSTEDS+MH        175     627   \n",
       "\n",
       "    LRT p-value  Bonferroni p-value Bonferroni significant          cAIC  \\\n",
       "1  3.268630e-11              0.0125                   True  47181.924003   \n",
       "2  1.935679e-03              0.0125                   True  47168.822954   \n",
       "3  9.103829e-15              0.0125                   True  46345.197711   \n",
       "4  1.262289e-02              0.0125                  False  46344.421537   \n",
       "\n",
       "         w1        p1  ...      SRV1    SRV_p1      SRV2    SRV_p2      SRV3  \\\n",
       "1  0.000000  0.241485  ...                                                     \n",
       "2  0.064925  0.928944  ...                                                     \n",
       "3  0.077807  0.891885  ...  0.497895  0.552027  1.340823   0.37766  3.111401   \n",
       "4  0.120585  0.993374  ...  0.499453   0.55085  1.335909  0.379458  3.127391   \n",
       "\n",
       "     SRV_p3 Convergence_Issue   DH_Rate   TH_Rate p_value_averaged  \n",
       "1                                                         0.007521  \n",
       "2                              0.044895  0.067012         0.007521  \n",
       "3  0.070314                                               0.007521  \n",
       "4  0.069692                    0.032234   0.14427         0.007521  \n",
       "\n",
       "[4 rows x 24 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "15857068",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BUSTED': {'weight': 1.3761003709062587e-182,\n",
       "  'original_pval': 3.268629811259416e-11,\n",
       "  'normalized_weighted': 8.199110165841534e-183},\n",
       " 'BUSTED+MH': {'weight': 9.627349162962485e-180,\n",
       "  'original_pval': 0.001935678907986405,\n",
       "  'normalized_weighted': 5.736187422154897e-180},\n",
       " 'BUSTEDS': {'weight': 0.6783533128256481,\n",
       "  'original_pval': 9.103828801926284e-15,\n",
       "  'normalized_weighted': 0.404177897253102},\n",
       " 'BUSTEDS+MH': {'weight': 1.0,\n",
       "  'original_pval': 0.01262288687330809,\n",
       "  'normalized_weighted': 0.5958221027468981}}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#for method in averaged.keys():\n",
    "#    averaged[method][\"normalized_weighted\"] = \n",
    "averaged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "amber-gregory",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-44-7560d6fdb3d2>:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_subset[\"Annotation\"][index] = \"Preferred\"\n"
     ]
    }
   ],
   "source": [
    "df_holder = []\n",
    "\n",
    "for g in set(result[\"Gene\"].to_list()):\n",
    "    #print(g)\n",
    "    df_subset = result[result[\"Gene\"] == g]\n",
    "    df_subset[\"Annotation\"] = \"\"\n",
    "    min_cAIC = min(df_subset[\"cAIC\"].to_list())\n",
    "    if len(df_subset[\"cAIC\"].to_list()) > 1:\n",
    "        second_smallest = sorted(df_subset[\"cAIC\"].to_list())[1]\n",
    "    else:\n",
    "        second_mallest = 0\n",
    "    #end if\n",
    "    for index, row in df_subset.iterrows():\n",
    "        caic = row[\"cAIC\"]\n",
    "        if caic == min_cAIC:\n",
    "            if second_smallest - caic > 5:\n",
    "                df_subset[\"Annotation\"][index] = \"Strongly Preferred\"\n",
    "            else:\n",
    "                df_subset[\"Annotation\"][index] = \"Preferred\"\n",
    "        #end if\n",
    "    #end for\n",
    "    df_holder.append(df_subset)\n",
    "#end for\n",
    "df_main = pd.concat(df_holder)\n",
    "df_main = df_main.reset_index(drop=True)\n",
    "df_main.index += 1\n",
    "#df_main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "brief-halloween",
   "metadata": {},
   "source": [
    "## Save table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "juvenile-depth",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving results to: mammalian_REM2_BUSTED_ModelTest.csv\n"
     ]
    }
   ],
   "source": [
    "#OUTPUT_CSV = \"Ebola.csv\"\n",
    "#OUTPUT_CSV = \"Zika.csv\"\n",
    "#OUTPUT_CSV = \"Monkeypox.csv\"\n",
    "#OUTPUT_CSV = TAG + \"-internal.csv\"\n",
    "\n",
    "print(\"Saving results to:\", OUTPUT_CSV)\n",
    "#result = result.T\n",
    "\n",
    "result.T.to_csv(os.path.join(\"..\",\n",
    "                           \"tables\",\n",
    "                           OUTPUT_CSV))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "premium-operation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Method</th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Codons</th>\n",
       "      <th>LRT p-value</th>\n",
       "      <th>Bonferroni p-value</th>\n",
       "      <th>Bonferroni significant</th>\n",
       "      <th>cAIC</th>\n",
       "      <th>w1</th>\n",
       "      <th>p1</th>\n",
       "      <th>...</th>\n",
       "      <th>SRV1</th>\n",
       "      <th>SRV_p1</th>\n",
       "      <th>SRV2</th>\n",
       "      <th>SRV_p2</th>\n",
       "      <th>SRV3</th>\n",
       "      <th>SRV_p3</th>\n",
       "      <th>Convergence_Issue</th>\n",
       "      <th>DH_Rate</th>\n",
       "      <th>TH_Rate</th>\n",
       "      <th>p_value_averaged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>3.268630e-11</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47181.924003</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.241485</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.007521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.935679e-03</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47168.822954</td>\n",
       "      <td>0.064925</td>\n",
       "      <td>0.928944</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.044895</td>\n",
       "      <td>0.067012</td>\n",
       "      <td>0.007521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>9.103829e-15</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>46345.197711</td>\n",
       "      <td>0.077807</td>\n",
       "      <td>0.891885</td>\n",
       "      <td>...</td>\n",
       "      <td>0.497895</td>\n",
       "      <td>0.552027</td>\n",
       "      <td>1.340823</td>\n",
       "      <td>0.37766</td>\n",
       "      <td>3.111401</td>\n",
       "      <td>0.070314</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.007521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.262289e-02</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>False</td>\n",
       "      <td>46344.421537</td>\n",
       "      <td>0.120585</td>\n",
       "      <td>0.993374</td>\n",
       "      <td>...</td>\n",
       "      <td>0.499453</td>\n",
       "      <td>0.55085</td>\n",
       "      <td>1.335909</td>\n",
       "      <td>0.379458</td>\n",
       "      <td>3.127391</td>\n",
       "      <td>0.069692</td>\n",
       "      <td></td>\n",
       "      <td>0.032234</td>\n",
       "      <td>0.14427</td>\n",
       "      <td>0.007521</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Gene      Method  Sequences  Codons  \\\n",
       "1  mammalian_REM2_codons.SA.FilterOutliers      BUSTED        175     627   \n",
       "2  mammalian_REM2_codons.SA.FilterOutliers   BUSTED+MH        175     627   \n",
       "3  mammalian_REM2_codons.SA.FilterOutliers     BUSTEDS        175     627   \n",
       "4  mammalian_REM2_codons.SA.FilterOutliers  BUSTEDS+MH        175     627   \n",
       "\n",
       "    LRT p-value  Bonferroni p-value Bonferroni significant          cAIC  \\\n",
       "1  3.268630e-11              0.0125                   True  47181.924003   \n",
       "2  1.935679e-03              0.0125                   True  47168.822954   \n",
       "3  9.103829e-15              0.0125                   True  46345.197711   \n",
       "4  1.262289e-02              0.0125                  False  46344.421537   \n",
       "\n",
       "         w1        p1  ...      SRV1    SRV_p1      SRV2    SRV_p2      SRV3  \\\n",
       "1  0.000000  0.241485  ...                                                     \n",
       "2  0.064925  0.928944  ...                                                     \n",
       "3  0.077807  0.891885  ...  0.497895  0.552027  1.340823   0.37766  3.111401   \n",
       "4  0.120585  0.993374  ...  0.499453   0.55085  1.335909  0.379458  3.127391   \n",
       "\n",
       "     SRV_p3 Convergence_Issue   DH_Rate   TH_Rate p_value_averaged  \n",
       "1                                                         0.007521  \n",
       "2                              0.044895  0.067012         0.007521  \n",
       "3  0.070314                                               0.007521  \n",
       "4  0.069692                    0.032234   0.14427         0.007521  \n",
       "\n",
       "[4 rows x 24 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0c0f1833",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = []\n",
    "for item in df_main[\"Gene\"].to_list():\n",
    "    gene_list.append(item.split(\"_\")[0])\n",
    "gene_set = set(gene_list)\n",
    "#gene_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4ea0114f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-29-f22282da0427>:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_main[\"Group\"][index] = str(gene.split(\"_\")[0])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Method</th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Codons</th>\n",
       "      <th>LRT p-value</th>\n",
       "      <th>Bonferroni p-value</th>\n",
       "      <th>Bonferroni significant</th>\n",
       "      <th>cAIC</th>\n",
       "      <th>w1</th>\n",
       "      <th>p1</th>\n",
       "      <th>...</th>\n",
       "      <th>SRV2</th>\n",
       "      <th>SRV_p2</th>\n",
       "      <th>SRV3</th>\n",
       "      <th>SRV_p3</th>\n",
       "      <th>Convergence_Issue</th>\n",
       "      <th>DH_Rate</th>\n",
       "      <th>TH_Rate</th>\n",
       "      <th>p_value_averaged</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>3.268630e-11</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47181.924003</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.241485</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.007521</td>\n",
       "      <td></td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTED+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.935679e-03</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>47168.822954</td>\n",
       "      <td>0.064925</td>\n",
       "      <td>0.928944</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.044895</td>\n",
       "      <td>0.067012</td>\n",
       "      <td>0.007521</td>\n",
       "      <td></td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>9.103829e-15</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>46345.197711</td>\n",
       "      <td>0.077807</td>\n",
       "      <td>0.891885</td>\n",
       "      <td>...</td>\n",
       "      <td>1.340823</td>\n",
       "      <td>0.37766</td>\n",
       "      <td>3.111401</td>\n",
       "      <td>0.070314</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.007521</td>\n",
       "      <td></td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mammalian_REM2_codons.SA.FilterOutliers</td>\n",
       "      <td>BUSTEDS+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>627</td>\n",
       "      <td>1.262289e-02</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>False</td>\n",
       "      <td>46344.421537</td>\n",
       "      <td>0.120585</td>\n",
       "      <td>0.993374</td>\n",
       "      <td>...</td>\n",
       "      <td>1.335909</td>\n",
       "      <td>0.379458</td>\n",
       "      <td>3.127391</td>\n",
       "      <td>0.069692</td>\n",
       "      <td></td>\n",
       "      <td>0.032234</td>\n",
       "      <td>0.14427</td>\n",
       "      <td>0.007521</td>\n",
       "      <td>Preferred</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>mammalian_REM2_codons.SA</td>\n",
       "      <td>BUSTED</td>\n",
       "      <td>175</td>\n",
       "      <td>656</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>53158.582997</td>\n",
       "      <td>0.037064</td>\n",
       "      <td>0.878519</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.470615</td>\n",
       "      <td></td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>mammalian_REM2_codons.SA</td>\n",
       "      <td>BUSTED+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>656</td>\n",
       "      <td>4.709140e-01</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>False</td>\n",
       "      <td>53018.100887</td>\n",
       "      <td>0.061488</td>\n",
       "      <td>0.568440</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.157208</td>\n",
       "      <td>1.268598</td>\n",
       "      <td>0.470615</td>\n",
       "      <td>Preferred</td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>mammalian_REM2_codons.SA</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>175</td>\n",
       "      <td>656</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>True</td>\n",
       "      <td>53158.779210</td>\n",
       "      <td>0.041782</td>\n",
       "      <td>0.886786</td>\n",
       "      <td>...</td>\n",
       "      <td>1.102266</td>\n",
       "      <td>0.429749</td>\n",
       "      <td>2.554419</td>\n",
       "      <td>0.129306</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.470615</td>\n",
       "      <td></td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>mammalian_REM2_codons.SA</td>\n",
       "      <td>BUSTEDS+MH</td>\n",
       "      <td>175</td>\n",
       "      <td>656</td>\n",
       "      <td>4.703152e-01</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>False</td>\n",
       "      <td>53018.101718</td>\n",
       "      <td>0.057268</td>\n",
       "      <td>0.122868</td>\n",
       "      <td>...</td>\n",
       "      <td>1.173861</td>\n",
       "      <td>0.401595</td>\n",
       "      <td>2.579359</td>\n",
       "      <td>0.113711</td>\n",
       "      <td></td>\n",
       "      <td>0.159088</td>\n",
       "      <td>1.272576</td>\n",
       "      <td>0.470615</td>\n",
       "      <td></td>\n",
       "      <td>mammalian</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Gene      Method  Sequences  Codons  \\\n",
       "1  mammalian_REM2_codons.SA.FilterOutliers      BUSTED        175     627   \n",
       "2  mammalian_REM2_codons.SA.FilterOutliers   BUSTED+MH        175     627   \n",
       "3  mammalian_REM2_codons.SA.FilterOutliers     BUSTEDS        175     627   \n",
       "4  mammalian_REM2_codons.SA.FilterOutliers  BUSTEDS+MH        175     627   \n",
       "5                 mammalian_REM2_codons.SA      BUSTED        175     656   \n",
       "6                 mammalian_REM2_codons.SA   BUSTED+MH        175     656   \n",
       "7                 mammalian_REM2_codons.SA     BUSTEDS        175     656   \n",
       "8                 mammalian_REM2_codons.SA  BUSTEDS+MH        175     656   \n",
       "\n",
       "    LRT p-value  Bonferroni p-value Bonferroni significant          cAIC  \\\n",
       "1  3.268630e-11              0.0125                   True  47181.924003   \n",
       "2  1.935679e-03              0.0125                   True  47168.822954   \n",
       "3  9.103829e-15              0.0125                   True  46345.197711   \n",
       "4  1.262289e-02              0.0125                  False  46344.421537   \n",
       "5  0.000000e+00              0.0125                   True  53158.582997   \n",
       "6  4.709140e-01              0.0125                  False  53018.100887   \n",
       "7  0.000000e+00              0.0125                   True  53158.779210   \n",
       "8  4.703152e-01              0.0125                  False  53018.101718   \n",
       "\n",
       "         w1        p1  ...      SRV2    SRV_p2      SRV3    SRV_p3  \\\n",
       "1  0.000000  0.241485  ...                                           \n",
       "2  0.064925  0.928944  ...                                           \n",
       "3  0.077807  0.891885  ...  1.340823   0.37766  3.111401  0.070314   \n",
       "4  0.120585  0.993374  ...  1.335909  0.379458  3.127391  0.069692   \n",
       "5  0.037064  0.878519  ...                                           \n",
       "6  0.061488  0.568440  ...                                           \n",
       "7  0.041782  0.886786  ...  1.102266  0.429749  2.554419  0.129306   \n",
       "8  0.057268  0.122868  ...  1.173861  0.401595  2.579359  0.113711   \n",
       "\n",
       "  Convergence_Issue   DH_Rate   TH_Rate p_value_averaged Annotation      Group  \n",
       "1                                               0.007521             mammalian  \n",
       "2                    0.044895  0.067012         0.007521             mammalian  \n",
       "3                                               0.007521             mammalian  \n",
       "4                    0.032234   0.14427         0.007521  Preferred  mammalian  \n",
       "5                                               0.470615             mammalian  \n",
       "6                    0.157208  1.268598         0.470615  Preferred  mammalian  \n",
       "7                                               0.470615             mammalian  \n",
       "8                    0.159088  1.272576         0.470615             mammalian  \n",
       "\n",
       "[8 rows x 26 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_main[\"Group\"] = \"\"\n",
    "for index, row in df_main.iterrows():\n",
    "    gene = row[\"Gene\"]\n",
    "    df_main[\"Group\"][index] = str(gene.split(\"_\")[0])\n",
    "df_main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e1fa996",
   "metadata": {},
   "source": [
    "## Visualize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "44cf86eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = result.rename(columns={\"LRT p-value\": \"p_value\"})\n",
    "df_main = df_main.rename(columns={\"LRT p-value\": \"p_value\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "014e93fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-6b31f57a4c3a44d58b403d85c7c7ee62\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-6b31f57a4c3a44d58b403d85c7c7ee62\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-6b31f57a4c3a44d58b403d85c7c7ee62\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelFontSize\": 12, \"labelLimit\": 1000, \"titleFontSize\": 12}, \"scale\": {\"bandPaddingInner\": 0.05}}, \"layer\": [{\"mark\": {\"type\": \"rect\", \"opacity\": 0.9}, \"encoding\": {\"color\": {\"condition\": {\"value\": \"darkblue\", \"test\": \"(datum.p_value <= 0.0125)\"}, \"value\": \"lightgray\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}, \"height\": 1200, \"width\": 600}, {\"mark\": \"text\", \"encoding\": {\"color\": {\"value\": \"white\"}, \"text\": {\"field\": \"Annotation\", \"type\": \"ordinal\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}, \"height\": 1200, \"width\": 600}], \"data\": {\"name\": \"data-0310b7ddbbc6a4bb5e3189fb5df01c2a\"}, \"title\": \"mammalian_REM2-InternalBranches Genome Scan with BUSTED, LRT pValue threshold: 0.0125\", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-0310b7ddbbc6a4bb5e3189fb5df01c2a\": [{\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 3.268629811259416e-11, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47181.92400308076, \"w1\": 0.0, \"p1\": 0.2414853979578538, \"w2\": 0.1474776619936387, \"p2\": 0.7540886032848153, \"w3\": 16.41081858920401, \"p3\": 0.004425998757330931, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.001935678907986405, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47168.82295423189, \"w1\": 0.06492506246216039, \"p1\": 0.9289439913708815, \"w2\": 1.0, \"p2\": 0.06963833871318946, \"w3\": 34.09655311705076, \"p3\": 0.001417669915929075, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": 0.04489548213047679, \"TH_Rate\": 0.06701152783406143, \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 9.103828801926284e-15, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 46345.1977106254, \"w1\": 0.0778070503003763, \"p1\": 0.8918851236751687, \"w2\": 0.5768919525352868, \"p2\": 0.1075503738088231, \"w3\": 196.8985592168631, \"p3\": 0.0005645025160082346, \"SRV1\": 0.4978947376614874, \"SRV_p1\": 0.5520267575582631, \"SRV2\": 1.340823177644006, \"SRV_p2\": 0.3776595345159393, \"SRV3\": 3.111400773474533, \"SRV_p3\": 0.07031370792579764, \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.01262288687330809, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 46344.42153659258, \"w1\": 0.1205852134638213, \"p1\": 0.9933739870258581, \"w2\": 1.0, \"p2\": 0.005928502293400546, \"w3\": 48.87740886791199, \"p3\": 0.0006975106807413471, \"SRV1\": 0.499453324667992, \"SRV_p1\": 0.5508499128657631, \"SRV2\": 1.335909358043514, \"SRV_p2\": 0.3794578603766473, \"SRV3\": 3.127391433823758, \"SRV_p3\": 0.0696922267575896, \"Convergence_Issue\": \"\", \"DH_Rate\": 0.03223415405438891, \"TH_Rate\": 0.1442702206979976, \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"Preferred\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTED\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.0, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 53158.5829970642, \"w1\": 0.03706405694542819, \"p1\": 0.8785192395679486, \"w2\": 1.0, \"p2\": 0.1176654756746579, \"w3\": 45.5801872561711, \"p3\": 0.003815284757393472, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTED+MH\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.4709139898870564, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 53018.10088731529, \"w1\": 0.06148845231279351, \"p1\": 0.5684398894222642, \"w2\": 0.09435576758906107, \"p2\": 0.3627879871282663, \"w3\": 1.216723154047895, \"p3\": 0.06877212344946951, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": 0.1572084465685608, \"TH_Rate\": 1.268598238662937, \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"Preferred\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTEDS\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.0, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 53158.779209679, \"w1\": 0.04178186337916327, \"p1\": 0.8867855884505349, \"w2\": 1.0, \"p2\": 0.1088409403512729, \"w3\": 38.45304497265218, \"p3\": 0.004373471198192249, \"SRV1\": 0.4445013850203849, \"SRV_p1\": 0.4409447824898033, \"SRV2\": 1.102265794666782, \"SRV_p2\": 0.4297492776246945, \"SRV3\": 2.554418650448586, \"SRV_p3\": 0.1293059398855022, \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTEDS+MH\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.4703152410983815, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 53018.10171764321, \"w1\": 0.05726811485115117, \"p1\": 0.122868311771324, \"w2\": 0.07215923357068714, \"p2\": 0.7977048132112093, \"w3\": 1.101235000047549, \"p3\": 0.07942687501746672, \"SRV1\": 0.4854219981186573, \"SRV_p1\": 0.4846936628503427, \"SRV2\": 1.173860595763572, \"SRV_p2\": 0.4015948843968688, \"SRV3\": 2.57935868683053, \"SRV_p3\": 0.1137114527527885, \"Convergence_Issue\": \"\", \"DH_Rate\": 0.1590881707685948, \"TH_Rate\": 1.272576459530427, \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"\", \"Group\": \"mammalian\"}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source = df_main\n",
    "pval_threshold = 0.05 / Tests\n",
    "\n",
    "heatmap = alt.Chart(source).mark_rect(opacity=0.9).encode(\n",
    "    y = alt.Y('Gene'),\n",
    "    x = alt.X('Method', axis=alt.Axis(labelAngle=-60)),\n",
    "    color = alt.condition((alt.datum.p_value <= pval_threshold), \n",
    "                                alt.ColorValue('darkblue'), \n",
    "                                alt.ColorValue('lightgray')) \n",
    ").properties(\n",
    "    width=600,\n",
    "    height=1200\n",
    ")\n",
    "\n",
    "text = heatmap.mark_text().encode(\n",
    "    text='Annotation:O',\n",
    "    color=alt.value('white'))\n",
    "\n",
    "chart = alt.layer(heatmap, text).configure_axis(\n",
    "    labelFontSize=12,\n",
    "    titleFontSize=12,\n",
    "    labelLimit = 1000\n",
    ").configure_scale(bandPaddingInner=0.05).properties(\n",
    "    title=TAG+\"-InternalBranches Genome Scan with BUSTED, LRT pValue threshold: \" + str(pval_threshold))\n",
    "\n",
    "chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "bee9914f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-f13f950736414953b86f3423b33eb89c\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-f13f950736414953b86f3423b33eb89c\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-f13f950736414953b86f3423b33eb89c\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelFontSize\": 12, \"labelLimit\": 1000, \"titleFontSize\": 12}, \"scale\": {\"bandPaddingInner\": 0.05}}, \"layer\": [{\"mark\": {\"type\": \"rect\", \"opacity\": 0.75}, \"encoding\": {\"color\": {\"condition\": {\"value\": \"darkblue\", \"test\": \"(datum.p_value_averaged <= 0.05)\"}, \"value\": \"gray\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}, \"height\": 1200, \"width\": 800}, {\"mark\": \"text\", \"encoding\": {\"color\": {\"value\": \"white\"}, \"text\": {\"field\": \"Annotation\", \"type\": \"ordinal\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}, \"height\": 1200, \"width\": 800}], \"data\": {\"name\": \"data-0310b7ddbbc6a4bb5e3189fb5df01c2a\"}, \"title\": \"mammalian_REM2 on Internal Branches Genome Scan with BUSTED, LRT p-value threshold: 0.05 on the AIC-weighted pvalue \", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-0310b7ddbbc6a4bb5e3189fb5df01c2a\": [{\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 3.268629811259416e-11, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47181.92400308076, \"w1\": 0.0, \"p1\": 0.2414853979578538, \"w2\": 0.1474776619936387, \"p2\": 0.7540886032848153, \"w3\": 16.41081858920401, \"p3\": 0.004425998757330931, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.001935678907986405, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47168.82295423189, \"w1\": 0.06492506246216039, \"p1\": 0.9289439913708815, \"w2\": 1.0, \"p2\": 0.06963833871318946, \"w3\": 34.09655311705076, \"p3\": 0.001417669915929075, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": 0.04489548213047679, \"TH_Rate\": 0.06701152783406143, \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 9.103828801926284e-15, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 46345.1977106254, \"w1\": 0.0778070503003763, \"p1\": 0.8918851236751687, \"w2\": 0.5768919525352868, \"p2\": 0.1075503738088231, \"w3\": 196.8985592168631, \"p3\": 0.0005645025160082346, \"SRV1\": 0.4978947376614874, \"SRV_p1\": 0.5520267575582631, \"SRV2\": 1.340823177644006, \"SRV_p2\": 0.3776595345159393, \"SRV3\": 3.111400773474533, \"SRV_p3\": 0.07031370792579764, \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.01262288687330809, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 46344.42153659258, \"w1\": 0.1205852134638213, \"p1\": 0.9933739870258581, \"w2\": 1.0, \"p2\": 0.005928502293400546, \"w3\": 48.87740886791199, \"p3\": 0.0006975106807413471, \"SRV1\": 0.499453324667992, \"SRV_p1\": 0.5508499128657631, \"SRV2\": 1.335909358043514, \"SRV_p2\": 0.3794578603766473, \"SRV3\": 3.127391433823758, \"SRV_p3\": 0.0696922267575896, \"Convergence_Issue\": \"\", \"DH_Rate\": 0.03223415405438891, \"TH_Rate\": 0.1442702206979976, \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"Preferred\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTED\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.0, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 53158.5829970642, \"w1\": 0.03706405694542819, \"p1\": 0.8785192395679486, \"w2\": 1.0, \"p2\": 0.1176654756746579, \"w3\": 45.5801872561711, \"p3\": 0.003815284757393472, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTED+MH\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.4709139898870564, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 53018.10088731529, \"w1\": 0.06148845231279351, \"p1\": 0.5684398894222642, \"w2\": 0.09435576758906107, \"p2\": 0.3627879871282663, \"w3\": 1.216723154047895, \"p3\": 0.06877212344946951, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": 0.1572084465685608, \"TH_Rate\": 1.268598238662937, \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"Preferred\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTEDS\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.0, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 53158.779209679, \"w1\": 0.04178186337916327, \"p1\": 0.8867855884505349, \"w2\": 1.0, \"p2\": 0.1088409403512729, \"w3\": 38.45304497265218, \"p3\": 0.004373471198192249, \"SRV1\": 0.4445013850203849, \"SRV_p1\": 0.4409447824898033, \"SRV2\": 1.102265794666782, \"SRV_p2\": 0.4297492776246945, \"SRV3\": 2.554418650448586, \"SRV_p3\": 0.1293059398855022, \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTEDS+MH\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.4703152410983815, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 53018.10171764321, \"w1\": 0.05726811485115117, \"p1\": 0.122868311771324, \"w2\": 0.07215923357068714, \"p2\": 0.7977048132112093, \"w3\": 1.101235000047549, \"p3\": 0.07942687501746672, \"SRV1\": 0.4854219981186573, \"SRV_p1\": 0.4846936628503427, \"SRV2\": 1.173860595763572, \"SRV_p2\": 0.4015948843968688, \"SRV3\": 2.57935868683053, \"SRV_p3\": 0.1137114527527885, \"Convergence_Issue\": \"\", \"DH_Rate\": 0.1590881707685948, \"TH_Rate\": 1.272576459530427, \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"\", \"Group\": \"mammalian\"}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source = df_main\n",
    "\n",
    "pval_threshold = 0.05\n",
    "\n",
    "heatmap = alt.Chart(source).mark_rect(opacity=0.75).encode(\n",
    "    y = alt.Y('Gene'),\n",
    "    x = alt.X('Method', axis=alt.Axis(labelAngle=-60)),\n",
    "    color = alt.condition((alt.datum.p_value_averaged <= pval_threshold), \n",
    "                                alt.ColorValue('darkblue'), \n",
    "                                alt.ColorValue('gray')) \n",
    ").properties(\n",
    "    width=800,\n",
    "    height=1200\n",
    ")\n",
    "\n",
    "text = heatmap.mark_text().encode(\n",
    "    text='Annotation:O',\n",
    "    color=alt.value('white'))\n",
    "\n",
    "chart = alt.layer(heatmap, text).configure_axis(\n",
    "    labelFontSize=12,\n",
    "    titleFontSize=12,\n",
    "    labelLimit = 1000\n",
    ").configure_scale(bandPaddingInner=0.05).properties(\n",
    "    title=TAG+\" on Internal Branches Genome Scan with BUSTED, LRT p-value threshold: \" + str(pval_threshold) + \" on the AIC-weighted pvalue \")\n",
    "\n",
    "chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0e045c1e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-77294ab99ec24339b82181eafe9d1368\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-77294ab99ec24339b82181eafe9d1368\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-77294ab99ec24339b82181eafe9d1368\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelFontSize\": 12, \"labelLimit\": 1000, \"titleFontSize\": 12}, \"scale\": {\"bandPaddingInner\": 0.05}}, \"layer\": [{\"mark\": {\"type\": \"rect\", \"opacity\": 0.75}, \"encoding\": {\"color\": {\"condition\": {\"value\": \"darkblue\", \"test\": \"(datum.p_value_averaged <= 0.05)\"}, \"value\": \"orange\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}, \"height\": 1200, \"width\": 800}, {\"mark\": \"text\", \"encoding\": {\"color\": {\"value\": \"white\"}, \"text\": {\"field\": \"Annotation\", \"type\": \"ordinal\"}, \"x\": {\"axis\": {\"labelAngle\": -60}, \"field\": \"Method\", \"type\": \"nominal\"}, \"y\": {\"field\": \"Gene\", \"type\": \"nominal\"}}, \"height\": 1200, \"width\": 800}], \"data\": {\"name\": \"data-0310b7ddbbc6a4bb5e3189fb5df01c2a\"}, \"encoding\": {}, \"title\": \"mammalian_REM2 on Internal Branches Genome Scan with BUSTED, LRT p-value threshold: 0.05 on the AIC-weighted pvalue \", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-0310b7ddbbc6a4bb5e3189fb5df01c2a\": [{\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 3.268629811259416e-11, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47181.92400308076, \"w1\": 0.0, \"p1\": 0.2414853979578538, \"w2\": 0.1474776619936387, \"p2\": 0.7540886032848153, \"w3\": 16.41081858920401, \"p3\": 0.004425998757330931, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTED+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.001935678907986405, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 47168.82295423189, \"w1\": 0.06492506246216039, \"p1\": 0.9289439913708815, \"w2\": 1.0, \"p2\": 0.06963833871318946, \"w3\": 34.09655311705076, \"p3\": 0.001417669915929075, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": 0.04489548213047679, \"TH_Rate\": 0.06701152783406143, \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 9.103828801926284e-15, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 46345.1977106254, \"w1\": 0.0778070503003763, \"p1\": 0.8918851236751687, \"w2\": 0.5768919525352868, \"p2\": 0.1075503738088231, \"w3\": 196.8985592168631, \"p3\": 0.0005645025160082346, \"SRV1\": 0.4978947376614874, \"SRV_p1\": 0.5520267575582631, \"SRV2\": 1.340823177644006, \"SRV_p2\": 0.3776595345159393, \"SRV3\": 3.111400773474533, \"SRV_p3\": 0.07031370792579764, \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA.FilterOutliers\", \"Method\": \"BUSTEDS+MH\", \"Sequences\": 175, \"Codons\": 627, \"p_value\": 0.01262288687330809, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 46344.42153659258, \"w1\": 0.1205852134638213, \"p1\": 0.9933739870258581, \"w2\": 1.0, \"p2\": 0.005928502293400546, \"w3\": 48.87740886791199, \"p3\": 0.0006975106807413471, \"SRV1\": 0.499453324667992, \"SRV_p1\": 0.5508499128657631, \"SRV2\": 1.335909358043514, \"SRV_p2\": 0.3794578603766473, \"SRV3\": 3.127391433823758, \"SRV_p3\": 0.0696922267575896, \"Convergence_Issue\": \"\", \"DH_Rate\": 0.03223415405438891, \"TH_Rate\": 0.1442702206979976, \"p_value_averaged\": 0.007520994999594324, \"Annotation\": \"Preferred\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTED\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.0, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 53158.5829970642, \"w1\": 0.03706405694542819, \"p1\": 0.8785192395679486, \"w2\": 1.0, \"p2\": 0.1176654756746579, \"w3\": 45.5801872561711, \"p3\": 0.003815284757393472, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTED+MH\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.4709139898870564, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 53018.10088731529, \"w1\": 0.06148845231279351, \"p1\": 0.5684398894222642, \"w2\": 0.09435576758906107, \"p2\": 0.3627879871282663, \"w3\": 1.216723154047895, \"p3\": 0.06877212344946951, \"SRV1\": \"\", \"SRV_p1\": \"\", \"SRV2\": \"\", \"SRV_p2\": \"\", \"SRV3\": \"\", \"SRV_p3\": \"\", \"Convergence_Issue\": \"\", \"DH_Rate\": 0.1572084465685608, \"TH_Rate\": 1.268598238662937, \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"Preferred\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTEDS\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.0, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"True\", \"cAIC\": 53158.779209679, \"w1\": 0.04178186337916327, \"p1\": 0.8867855884505349, \"w2\": 1.0, \"p2\": 0.1088409403512729, \"w3\": 38.45304497265218, \"p3\": 0.004373471198192249, \"SRV1\": 0.4445013850203849, \"SRV_p1\": 0.4409447824898033, \"SRV2\": 1.102265794666782, \"SRV_p2\": 0.4297492776246945, \"SRV3\": 2.554418650448586, \"SRV_p3\": 0.1293059398855022, \"Convergence_Issue\": \"\", \"DH_Rate\": \"\", \"TH_Rate\": \"\", \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"\", \"Group\": \"mammalian\"}, {\"Gene\": \"mammalian_REM2_codons.SA\", \"Method\": \"BUSTEDS+MH\", \"Sequences\": 175, \"Codons\": 656, \"p_value\": 0.4703152410983815, \"Bonferroni p-value\": 0.0125, \"Bonferroni significant\": \"False\", \"cAIC\": 53018.10171764321, \"w1\": 0.05726811485115117, \"p1\": 0.122868311771324, \"w2\": 0.07215923357068714, \"p2\": 0.7977048132112093, \"w3\": 1.101235000047549, \"p3\": 0.07942687501746672, \"SRV1\": 0.4854219981186573, \"SRV_p1\": 0.4846936628503427, \"SRV2\": 1.173860595763572, \"SRV_p2\": 0.4015948843968688, \"SRV3\": 2.57935868683053, \"SRV_p3\": 0.1137114527527885, \"Convergence_Issue\": \"\", \"DH_Rate\": 0.1590881707685948, \"TH_Rate\": 1.272576459530427, \"p_value_averaged\": 0.4706146776374474, \"Annotation\": \"\", \"Group\": \"mammalian\"}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source = df_main.copy()\n",
    "\n",
    "pval_threshold = 0.05\n",
    "\n",
    "heatmap = alt.Chart(source).mark_rect(opacity=0.75).encode(\n",
    "    y = alt.Y('Gene'),\n",
    "    x = alt.X('Method', axis=alt.Axis(labelAngle=-60)),\n",
    "    color = alt.condition((alt.datum.p_value_averaged <= pval_threshold), \n",
    "                                alt.ColorValue('darkblue'), \n",
    "                                alt.ColorValue('orange')),\n",
    ").properties(\n",
    "    width=800,\n",
    "    height=1200\n",
    ")\n",
    "\n",
    "text = heatmap.mark_text().encode(\n",
    "    text='Annotation:O',\n",
    "    color=alt.value('white'))\n",
    "\n",
    "chart = alt.layer(heatmap, text).encode(\n",
    ").configure_axis(\n",
    "    labelFontSize=12,\n",
    "    titleFontSize=12,\n",
    "    labelLimit = 1000,\n",
    ").configure_scale(bandPaddingInner=0.05).properties(\n",
    "    title=TAG+\" on Internal Branches Genome Scan with BUSTED, LRT p-value threshold: \" + str(pval_threshold) + \" on the AIC-weighted pvalue \")\n",
    "\n",
    "chart"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
